// @sap/cds-compiler/lib/render/renderUtil.js
// Common render functions for toCdl.js and toSql.js

const functionsWithoutParams = {
  hana: {
    CURRENT_CONNECTION: {},
    CURRENT_SCHEMA: {},
    CURRENT_TRANSACTION_ISOLATION_LEVEL: {},
    CURRENT_UTCDATE: {},
    CURRENT_UTCTIME: {},
    CURRENT_UTCTIMESTAMP: {},
    SYSUUID: {},
  }
}

// Dialect = 'hana' (only relevance at the moment) | 'cap' | 'sqlite'
function renderFunc( node, dialect, renderArgs, parenNameToUpper = false )  {
  if (funcWithoutParen( node, dialect ))
    return node.func;
  else
    // unclear why we transform the function name to uppercase in SQL, but only with parentheses
    return `${parenNameToUpper ? node.func.toUpperCase() : node.func}(${renderArgs( node.args )})`;
}

function funcWithoutParen( node, dialect ) {
  if (!node.args)
    return true;
  if (!Array.isArray( node.args ) || node.args.length)
    return false;
  const specials = functionsWithoutParams[dialect];
  return specials && specials[node.func.toUpperCase()];
}


module.exports = {
  renderFunc
}
