// @sap/hana-client/extension/TypeCode.js
'use strict';

module.exports = {
    NULL: 0,
    TINYINT: 1,
    SMALLINT: 2,
    INT: 3,
    BIGINT: 4,
    DECIMAL: 5,
    REAL: 6,
    DOUBLE: 7,
    CHAR: 8,
    VARCHAR1: 9,
    NCHAR: 10,
    NVARCHAR: 11,
    BINARY: 12,
    VARBINARY: 13,
    DATE: 14,
    TIME: 15,
    TIMESTAMP: 16,
    TIME_TZ: 17,
    TIME_LTZ: 18,
    TIMESTAMP_TZ: 19,
    TIMESTAMP_LTZ: 20,
    INTERVAL_YM: 21,
    INTERVAL_DS: 22,
    ROWID: 23,
    UROWID: 24,
    CLOB: 25,
    NCLOB: 26,
    BLOB: 27,
    BOOLEAN: 28,
    STRING: 29,
    NSTRING: 30,
    LOCATOR: 31,
    NLOCATOR: 32,
    BSTRING: 33,
    DECIMAL_DIGIT_ARRAY: 34,
    VARCHAR2: 35,
    UNUSED1: 36,
    UNUSED2: 37,
    UNUSED3: 38,
    UNUSED4: 39,
    UNUSED5: 40,
    UNUSED6: 41,
    UNUSED7: 42,
    UNUSED8: 43,
    UNUSED9: 44,
    TABLE: 45,
    UNUSED11: 46,
    UNUSED1F: 47,
    ABAPSTREAM: 48,
    ABAPSTRUCT: 49,
    UNUSED12: 50,
    TEXT: 51,
    SHORTTEXT: 52,
    BINTEXT: 53,
    UNUSED16: 54,
    ALPHANUM: 55,
    UNUSED18: 56,
    UNUSED19: 57,
    UNUSED20: 58,
    UNUSED21: 59,
    UNUSED22: 60,
    LONGDATE: 61,
    SECONDDATE: 62,
    DAYDATE: 63,
    SECONDTIME: 64,
    CSDATE: 65,
    CSTIME: 66,
    BLOB_DISK: 71,
    CLOB_DISK: 72,
    NCLOB_DISK: 73
};
