// @sap/cds-runtime/lib/cds-services/adapter/rest/utils/parse-url.js
const getError = require('../../../util/getError')
const { getConvertedValue } = require('./key-value-utils')
const { checkStatic } = require('../../../util/assert')

const _normalizeAndSplitUrl = req => {
  // Normalize /path////to/someWhere into path/to/someWhere and split by path
  const parts = `/${req.path}/`
    .replace(/[/]+/g, '/')
    .replace(/\/(.*)\//, '$1')
    .split('/')

  // TODO: replace with generic solution, currently we do not support longer urls
  if (parts.length > 3) {
    throw getError(400, 'navigation is not yet supported')
  }

  return parts
}

const _enrichCustomOperation = (csnElement, customOperationName) => {
  return Object.assign({ name: customOperationName }, csnElement)
}

const _initializeParsed = method => {
  return {
    method: method,
    segments: []
  }
}

const _parseEntityOrOperation = part => {
  let decodedPart = decodeURI(part)
  decodedPart = decodedPart.replace(/"/gi, '')
  const [, name, paramsString = ''] = decodedPart.match(/([^(]+)\(?(.*[^)]+)?\)?/)
  const params = paramsString
    .split(',')
    .map(keyValue => keyValue.split('='))
    .reduce((obj, [key, value]) => {
      if (key) {
        obj[key] = value
      }
      return obj
    }, {})
  const returnObj = { name }
  if (Object.keys(params).length > 0) {
    returnObj.params = params
  }
  return returnObj
}

const _findEntityOrCustomOperation = (customOperation, service, name) => {
  const thing = service.entities[name] || service.operations[name]
  if (!thing) {
    throw getError(404, `Requested resource ${name} is not found`)
  }

  if (thing.kind === 'entity') {
    return thing
  }

  if (thing.kind === customOperation) {
    return _enrichCustomOperation(thing, name)
  }

  throw getError(400, `${name} is neither a valid resource or an unbound ${customOperation}`)
}

const _validateEntity = entity => {
  if (!entity) {
    throw getError(404)
  }

  if (entity.kind === 'entity') {
    return entity
  }

  throw getError(400, `${entity.name} is not a valid resource`)
}

const _validateCustomOperation = (entity, name, customOperation) => {
  if (entity.actions && entity.actions[name] && entity.actions[name].kind === customOperation) {
    return _enrichCustomOperation(entity.actions[name], name)
  }

  throw getError(400, `${name} is not a bound ${customOperation}`)
}

const _validateAndConvertParamValues = (csnElement, params = {}) => {
  for (const param of Object.keys(params)) {
    const csnElementParam = csnElement.params[param]
    if (!csnElementParam) {
      throw getError(400, `Invalid parameter: ${param}`)
    }
    const convertedParam = getConvertedValue(csnElementParam.type, params[param])
    if (Number.isNaN(convertedParam)) {
      throw getError(400, `Parameter value for '${param}' must be of type ${csnElementParam.type}`)
    }
    params[param] = convertedParam
  }
  checkStatic({ elements: csnElement.params }, params)
}

const _setConvenienceProperties = parsed => {
  const lastElement = parsed.segments[parsed.segments.length - 1]

  if (typeof lastElement === 'string') {
    parsed.isCollection = false
  } else if (lastElement.kind === 'entity') {
    parsed.isCollection = true
  } else {
    parsed.customOperation = parsed.segments.length === 1 ? `unbound${lastElement.kind}` : `bound${lastElement.kind}`
  }
}

const _parseCreateOrRead1 = (parts, customOperation, service, parsed) => {
  const { name, params } = _parseEntityOrOperation(parts[0])
  const entityOrCustomOperation = _findEntityOrCustomOperation(customOperation, service, name)
  if (params) {
    _validateAndConvertParamValues(entityOrCustomOperation, params)
  }
  if (params && customOperation === 'function') {
    parsed.params = params
  }
  parsed.segments.push(entityOrCustomOperation)
}

const _parseCreateOrRead2 = (method, parsed, service, parts) => {
  if (method === 'CREATE') {
    throw getError(400, 'POST is only supported on resource collections or actions')
  }
  parsed.segments.push(_validateEntity(service.entities[parts[0]]), parts[1])
}

const _parseCreateOrRead3 = (service, parts, customOperation, parsed) => {
  const entity = _validateEntity(service.entities[parts[0]])
  const key = parts[1]
  const { name, params } = _parseEntityOrOperation(parts[2])
  const operation = _validateCustomOperation(entity, name, customOperation)
  if (params) {
    _validateAndConvertParamValues(operation, params)
  }
  if (params && customOperation === 'function') {
    parsed.params = params
  }
  parsed.segments.push(entity, key, operation)
}

const parseCreateOrReadUrl = (method, service, req) => {
  const parts = _normalizeAndSplitUrl(req)
  const customOperation = method === 'READ' ? 'function' : 'action'
  const parsed = _initializeParsed(method)

  // TODO: replace with generic solution
  if (parts.length === 1) {
    _parseCreateOrRead1(parts, customOperation, service, parsed)
  }

  if (parts.length === 2) {
    _parseCreateOrRead2(method, parsed, service, parts)
  }

  if (parts.length === 3) {
    _parseCreateOrRead3(service, parts, customOperation, parsed)
  }

  _setConvenienceProperties(parsed)

  return parsed
}

const parseUpdateOrDeleteUrl = (method, entities, req) => {
  const parts = _normalizeAndSplitUrl(req)

  // TODO: replace with generic solution
  if (method === 'DELETE' && parts.length !== 2) {
    throw getError(400, 'DELETE is only supported on a specific resource')
  }

  if (method === 'UPDATE' && parts.length !== 1 && parts.length !== 2) {
    throw getError(400, 'PUT/PATCH is only supported on a specific resource or resource collection')
  }

  const entity = _validateEntity(entities[parts[0]])
  const segments = [entity]

  if (parts[1]) {
    segments.push(parts[1])
  }

  return {
    method,
    segments
  }
}

module.exports = {
  parseCreateUrl: (service, req) => {
    return parseCreateOrReadUrl('CREATE', service, req)
  },
  parseReadUrl: (service, req) => {
    return parseCreateOrReadUrl('READ', service, req)
  },
  parseUpdateUrl: (entities, req) => {
    return parseUpdateOrDeleteUrl('UPDATE', entities, req)
  },
  parseDeleteUrl: (entities, req) => {
    return parseUpdateOrDeleteUrl('DELETE', entities, req)
  }
}
