// @sap/cds/lib/build/node-cf/index.js
const fs = require('@sap/cds-foss')('fs-extra')
const path = require('path')
const BuildTaskHandlerOData = require('../buildTaskHandlerOData')
const { impl4 } = require("../../srv/serve")
const { getApplicationName, getHdiService, getDefaultHdiServiceName } = require('../../utils/mtaUtil')
const { BuildMessage } = require('../util')
const { BUILD_OPTION_OUTPUT_MODE, OUTPUT_MODE_RESULT_ONLY, ODATA_VERSION, ODATA_VERSION_V2,
    BUILD_NODEJS_EDMX_GENERAION, BUILD_TASK_HANA, FOLDER_GEN, SEVERITY_WARNING } = require('../constants')

const DEBUG = process.env.DEBUG
const FILE_NAME_MANIFEST_YAML = "manifest.yaml"

class NodeCfModuleBuilder extends BuildTaskHandlerOData {
    constructor(task, buildOptions) {
        super("Node CF Module Builder", task, buildOptions)

        if (this.isStagingBuild()) {
            const srcDir = path.relative(this.buildOptions.root, this.task.src)
            // fallback if src has been defined as '.'
            this.task.dest = path.resolve(this.task.dest, srcDir || "srv")
        }
    }

    async build() {
        // resolve specified models, using all models by default, e.g.
        // { use:'...', src:'srv', options:{model:['app','srv']} }
        // { use:'...', src:'srv' }
        const { src, dest } = this.task
        const modelPaths = this.resolveModel()
        const destGen = this.isStagingBuild() ? dest : path.join(dest, FOLDER_GEN)

        this.logger.log(`\n[cds] - building module [${this.stripProjectPaths(src)}] using [${this.name}]`)

        if (this.hasCdsEnvOption(ODATA_VERSION, ODATA_VERSION_V2)) {
            // log warning as nodejs is only supporting odata version V4
            this.pushMessages(new BuildMessage("odata.version = 'v2' not supported by node runtime, define odata.version = 'v4' in package.json instead", SEVERITY_WARNING))
        }

        if (!modelPaths || modelPaths.length === 0) {
            this.logger.log("[cds] - no model found, skip build")
            return this._result
        }
        if (DEBUG) {
            this.logger.log(`[cds] - model: ${this.stripProjectPaths(modelPaths).join(", ")}`)
        }

        // compile to csn
        const options = this.options()
        const model = await this.cds.load(modelPaths, options)
        this.pushMessages(options.messages || options._messages)

        const csn = await this._compileCsn(modelPaths, destGen)

        // The option OUTPUT_MODE_RESULT_ONLY is set by mtx build_helper and used in this context to indicate that validation
        // of custom service handlers must be skiped as part of onboarding build as the custom handler functions are not part
        // of the extension respectively tenant - otherwise false possitive warnings would be logged.
        if (!this.hasBuildOption(BUILD_OPTION_OUTPUT_MODE, OUTPUT_MODE_RESULT_ONLY)) {
            this._validateCustomServiceHandlers(csn)
        }

        const promises = []
        const bundles = this.collectBundles(csn._sources)
        const {
            folders = ['i18n'], file = 'i18n'
        } = this.env.i18n

        if (this.hasCdsEnvOption(BUILD_NODEJS_EDMX_GENERAION, true) || this.hasBuildOption('edmxGeneration', true)) {
            promises.push(this.compileEdmx(model, destGen))
        }

        if (!this.hasBuildOption(BUILD_OPTION_OUTPUT_MODE, OUTPUT_MODE_RESULT_ONLY)) {
            promises.push(this._copyNativeContent(src, dest))

            if (!this.task.options.skipManifestGeneration) {
                promises.push(this._writeManifestYaml(), this._writeCfIgnore())
            }

            if (Object.keys(bundles).length > 0) {
                promises.push(this.write(bundles).to(path.join(destGen, folders[0], file + '.json')))
            }
            await Promise.all(promises)
        }
        return this._result
    }

    async clean() {
        // in staging build we want to delete the entire 'task.dest' dir
        // see #constructor - the 'src' folder has been appended to the origin 'task.dest' dir
        let genDest = this.isStagingBuild() ? path.dirname(this.task.dest) : path.join(this.task.dest, FOLDER_GEN)

        if (DEBUG) {
            this.logger.log(`Deleting build target folder ${genDest}`)
        }
        await fs.remove(genDest)
    }

    async _compileCsn(modelPaths, csnDest) {
        // REVISIT: Wieso verwendet ihr hier private cdsv API? -> No Go!
        const cdsv = require('../../models/cdsv')
        const csn = await cdsv.compile(modelPaths, this.buildOptions.root)

        if (!this.hasBuildOption(BUILD_OPTION_OUTPUT_MODE, OUTPUT_MODE_RESULT_ONLY)) {
            const json = this.cds.compile.to.json(csn)  //> @source wird erst in cds.compile.to.json hinzugefügt
            await this.write(json).to(path.join(csnDest, "csn.json"))
        }
        this._result.csn = csn

        return csn
    }

    async _copyNativeContent(src, dest) {
        await super.copyNativeContent(src, dest, (entry) => {
            if (fs.statSync(entry).isDirectory()) {
                return true // using common filter for folders
            }
            return !/\.cds$|package\.json$|manifest\.y.?ml$/.test(entry)
        })

        // copy relevant content located at project root, e.g. *.js, package.json -> gen/srv
        await super.copyNativeContent(this.buildOptions.root, path.dirname(dest), (entry) => {
            if (fs.statSync(entry).isDirectory()) {
                return path.basename(entry) === "handlers"
            }
            const baseName = path.basename(entry)
            if (/\.js$|^package\.json/.test(baseName)) {
                return baseName !== "package.json" || !this.task.options.skipPackageJsonGeneration
            }
            return false
        })
    }

    // this is to have a dev fast-turnaround
    async _writeCfIgnore() {
        if (this.isStagingBuild()) {
            const content = `node_modules/\n`
            await this.write(content).to(path.join(path.dirname(this.task.dest), '.cfignore'))
        }
    }

    async _writeManifestYaml() {
        if (!this.isStagingBuild()) {
            return
        }

        let manifest = path.join(this.task.src, FILE_NAME_MANIFEST_YAML)
        if (await fs.pathExists(manifest)) {
            // copy existing manifest to service root folder in staging area
            this.copy(manifest).to(path.join(path.dirname(this.task.dest), FILE_NAME_MANIFEST_YAML))
            return
        }
        manifest = path.join(this.task.src, 'manifest.yml')
        if (await fs.pathExists(manifest)) {
            // copy existing manifest to service root folder in staging area
            this.copy(manifest).to(path.join(path.dirname(this.task.dest), 'manifest.yml'))
            return
        }

        // generate one...
        // check whether a hdi service binding is required
        const hanaBuildTask = this.buildOptions.tasks.find(task => task.for === BUILD_TASK_HANA)
        if (!hanaBuildTask) {
            if (DEBUG) {
                this.logger.log("[cds] - skip creating [manifest.yaml] - no [hana] build task existing")
            }
            return
        }

        try {
            const [hdiService, applicationName] = await Promise.all([
                getHdiService(this.buildOptions.root, hanaBuildTask.src, this.logger),
                getApplicationName(this.buildOptions.root, this.task.src, "nodejs", this.logger)
            ])

            const hdiServiceName = hdiService ? hdiService.name : getDefaultHdiServiceName(this.buildOptions.root, hanaBuildTask.src);
            const MANIFEST_YAML_CONTENT = `---
applications:
- name: ${applicationName}
  path: .
  memory: 128M
  services:
  - ${hdiServiceName}`

            await this.write(MANIFEST_YAML_CONTENT).to(path.join(path.dirname(this.task.dest), FILE_NAME_MANIFEST_YAML))
        } catch (e) {
            if (e.name === 'YAMLSyntaxError') {
                this.logger.error("Failed to parse [mta.yaml]")
            }
            this.logger.error(e)
        }
    }

    /**
     * Validate custom service handler defined using @impl annotation.
     * Issue warning in case the handler function cannot be resolved.
     * @param {object} csn
     */
    _validateCustomServiceHandlers(csn) {
        const _exclude = (def) => def['@cds.serve.ignore'] || def['@cds.ignore'] || def.kind !== "service" || !def['@impl']
        this.cds.reflect(csn).foreach((def) => {
            if (!_exclude(def)) {
                try {
                    impl4(def, def['@impl'], this.task.src)
                } catch (e) {
                    this.pushMessages(new BuildMessage(`Failed to resolve custom service handler. ${e.message}`, SEVERITY_WARNING))
                }
            }
        })
    }
}

module.exports = NodeCfModuleBuilder
