// @sap/cds-runtime/lib/cds-services/services/handlers/index.js
/**
 * @module handlers
 */
module.exports = {
  afterCommitPersonalData: require('./afterCommitPersonalData'),
  afterRollbackPersonalData: require('./afterRollbackPersonalData'),
  beforeETagCheck: require('../../../common/generic/etag').handler,
  beforeCommitPersonalData: require('./beforeCommitPersonalData'),
  beforeCreateDraft: require('./beforeCreateDraft'),
  beforeDeleteOrCancelDraft: require('./beforeDeleteOrCancelDraft'),
  beforeEntityAuth: require('./beforeEntityAuth'),
  beforeRollbackPersonalData: require('./beforeRollbackPersonalData'),
  beforeUpdateDraft: require('./beforeUpdateDraft'),
  beforeDefaultValuesKeys: require('./beforeDefaultValuesKeys'),
  beforeInputValidation: require('./beforeInputValidation'),
  finalOn: require('./finalOn'),
  onCommit: require('./onCommit'),
  onCreate: require('./onCreate'),
  onCreateDraft: require('./onCreateDraft'),
  onDelete: require('./onDelete'),
  onDeleteDraft: require('./onDeleteDraft'),
  onCancelDraft: require('./onCancelDraft'),
  onDraftActivate: require('./onDraftActivate'),
  onDraftActivateEvent: require('./onDraftActivateEvent'),
  onDraftEdit: require('./onDraftEdit'),
  onDraftPrepare: require('./onDraftPrepare'),
  onRead: require('./onRead'),
  onReadDraft: require('./onReadDraft'),
  onReadOverDraft: require('./onReadOverDraft'),
  onRollback: require('./onRollback'),
  onUpdate: require('./onUpdate'),
  onPatchDraft: require('./onPatchDraft')
}
