// @sap/cds-runtime/lib/cds-services/services/handlers/onCreateDraft.js
const onDraftActivate = require('./onDraftActivate')()
const { isNavigationToMany } = require('../utils/compositionTree')
const { removeDraftUUID, ensureDraftsSuffix } = require('../utils/draftUtils')
const {
  messages: { DB_CONNECTION_MISSING },
  DRAFT_COLUMNS
} = require('../utils/constants')

const _getUpdateDraftAdminCQN = ({ statements, user }, draftUUID) => {
  return statements
    .UPDATE('DRAFT.DraftAdministrativeData')
    .data({
      InProcessByUser: user.id,
      LastChangedByUser: user.id,
      LastChangeDateTime: new Date().toISOString()
    })
    .where({ DraftUUID: draftUUID })
}

const _getInsertDraftAdminCQN = ({ statements, user }, uuid) => {
  const time = new Date().toISOString()

  return statements.INSERT.into('DRAFT.DraftAdministrativeData').entries({
    DraftUUID: uuid,
    CreationDateTime: time,
    CreatedByUser: user.id,
    LastChangeDateTime: time,
    LastChangedByUser: user.id,
    DraftIsCreatedByMe: true,
    DraftIsProcessedByMe: true,
    InProcessByUser: user.id
  })
}

const _getInsertDataCQN = (context, draftUUID) => {
  const draftName = ensureDraftsSuffix(context.target.name)

  const insertData = context.statements.INSERT.into(draftName).entries(context.query.INSERT.entries[0]) // entries is always set because there are no entities without keys

  context.data.IsActiveEntity = false
  context.data.HasDraftEntity = false
  context.data.HasActiveEntity = false
  context.data.DraftAdministrativeData_DraftUUID = draftUUID

  return insertData
}

/**
 * Generic Handler for CREATE requests in the context of draft.
 * In case of success it returns the created entry.
 *
 * @alias module:handlers.onCreateDraft
 */
const onCreateDraft = () => context => {
  if (!context.isDraftChange) {
    return onDraftActivate(context)
  }

  if (!context.run) {
    context.log.warn(DB_CONNECTION_MISSING)
    return Promise.resolve(context.data)
  }

  // fill default values
  const elements = context.target.elements
  for (const column of Object.keys(elements)) {
    const col = elements[column]
    if (col.default !== undefined && !DRAFT_COLUMNS.includes(column)) {
      context.data[col.name] = 'val' in col.default ? col.default.val : col.default
    }
  }

  const navigationToMany = isNavigationToMany(context)

  const adminDataCQN = navigationToMany
    ? _getUpdateDraftAdminCQN(context, context.data.DraftAdministrativeData_DraftUUID)
    : _getInsertDraftAdminCQN(context, context.data.DraftAdministrativeData_DraftUUID)
  const insertDataCQN = _getInsertDataCQN(context, context.data.DraftAdministrativeData_DraftUUID)

  // read data as on db and return
  const keyName = Object.values(context.target.elements).find(val => val.key).name
  const readInsertDataCQN = context.statements.SELECT.from(insertDataCQN.INSERT.into).where(
    keyName,
    '=',
    context.data[keyName]
  )

  return Promise.all([context.run(adminDataCQN), context.run(insertDataCQN)])
    .then(() => {
      return context.run(readInsertDataCQN)
    })
    .then(result => {
      if (result.length === 0) {
        context.reject(404)
      }
      return removeDraftUUID(result[0])
    })
}

module.exports = onCreateDraft
