// @sap/cds-runtime/lib/cds-services/services/handlers/onDraftEdit.js
const generateUUID = require('@sap/cds-foss')('uuid/v4')
const { getCompositionTree } = require('../utils/compositionTree')
const getColumns = require('../utils/columns')
const {
  draftIsLocked,
  ensureDraftsSuffix,
  ensureNoDraftsSuffix,
  getDraftName,
  getEntityName,
  getServiceNamespace,
  getSubCQNs,
  setStatusCodeAndHeader
} = require('../utils/draftUtils')
const { isActiveEntityRequested, getKeyData } = require('../utils/draftWhereUtils')

const {
  messages: { DB_CONNECTION_MISSING }
} = require('../utils/constants')

const _getDraftColumns = draftUUID => ({
  IsActiveEntity: false,
  HasDraftEntity: false,
  HasActiveEntity: true,
  DraftAdministrativeData_DraftUUID: draftUUID
})

const _getAdminData = ({ user }, draftUUID, time) => {
  const currentUser = user.id || null
  return {
    DraftUUID: draftUUID,
    CreationDateTime: time,
    CreatedByUser: currentUser,
    LastChangeDateTime: time,
    LastChangedByUser: currentUser,
    DraftIsCreatedByMe: true,
    DraftIsProcessedByMe: true,
    InProcessByUser: currentUser
  }
}

const _getInsertAdminDataCQN = ({ statements: { INSERT }, user }, draftUUID, time) => {
  return INSERT.into('DRAFT.DraftAdministrativeData').entries(_getAdminData({ user }, draftUUID, time))
}

/**
 * Generic Handler for draftEdit requests.
 * In case of success it returns the root draft entry.
 *
 * @param context - operation object, that provides error, continuation and other functions as well as information
 * regarding the current operation.
 * @alias module:handlers.onDraftEdit
 */
const onDraftEdit = ({ model: { definitions } = {} } = {}) => async context => {
  if (!isActiveEntityRequested(context.query.SELECT.where || [])) {
    context.reject(400)
    return
  }

  if (!context.run) {
    context.log.warn(DB_CONNECTION_MISSING)
    context.reject(404)
    return
  }

  // TODO replace with generic where filter
  const keys = Object.keys(context.target.keys).filter(key => key !== 'IsActiveEntity')
  const data = getKeyData(keys, context.query.SELECT.where)
  const rootWhere = keys.reduce((res, key) => {
    res[key] = data[key]
    return res
  }, {})

  const rootCQN = context.statements.SELECT.from(context.target, getColumns(context.target, true)).where(rootWhere)

  const subCQNs = getSubCQNs({
    definitions,
    context,
    rootCQN,
    compositionTree: getCompositionTree(definitions, ensureNoDraftsSuffix(context.target.name))
  })

  const selectCQNs = [rootCQN, ...subCQNs.map(obj => obj.cqn)]

  const results = await Promise.all(selectCQNs.map(CQN => context.run(CQN)))
  if (!results[0].length) {
    context.reject(404)
    return
  }

  const serviceNamespace = getServiceNamespace(ensureNoDraftsSuffix(context.target.name))
  const entityName = getEntityName(ensureNoDraftsSuffix(context.target.name))
  const rootDraftName = ensureDraftsSuffix(context.target.name)

  const draftExists = await context.run(
    context.statements.SELECT.from(rootDraftName, ['DraftAdministrativeData_DraftUUID as DraftUUID']).where(rootWhere)
  )
  if (draftExists.length) {
    const adminData = await context.run(
      context.statements.SELECT.one('DRAFT.DraftAdministrativeData', ['InProcessByUser', 'LastChangeDateTime']).where(
        draftExists[0]
      )
    )

    // draft is locked OR draft is not locked but must be rejected for popup
    if (
      draftIsLocked(adminData.LastChangeDateTime) ||
      (!draftIsLocked(adminData.LastChangeDateTime) && context.data.PreserveChanges)
    ) {
      context.reject(409, 'Draft for object already exists')
      return
    }

    await Promise.all([
      context.run(context.statements.DELETE.from('DRAFT.DraftAdministrativeData').where(draftExists[0])),
      context.run(context.statements.DELETE.from(rootDraftName).where(rootWhere))
    ])
  }

  const time = new Date().toISOString()
  const draftUUID = generateUUID()
  const insertCQNs = [_getInsertAdminDataCQN(context, draftUUID, time)]

  for (const resultIndex in results) {
    if (results[resultIndex].length === 0) continue
    const draftEntity = getDraftName(serviceNamespace, getEntityName(selectCQNs[resultIndex].SELECT.from.ref[0]))
    const entries = results[resultIndex].map(entityResult =>
      Object.assign({}, entityResult, _getDraftColumns(draftUUID))
    )

    insertCQNs.push(context.statements.INSERT.into(draftEntity).entries(entries))
  }

  await Promise.all(insertCQNs.map(CQN => context.run(CQN)))
  setStatusCodeAndHeader(context._.odataRes, rootWhere, entityName, false)

  return Object.assign({}, results[0][0], {
    HasDraftEntity: false,
    HasActiveEntity: true,
    IsActiveEntity: false
  })
}

module.exports = onDraftEdit
