// @sap/odata-server/lib/core/OdataResponseInBatch.js
'use strict';

const SerializationError = require('../errors/SerializationError');

/**
 * Response object wrapper to carry original response information.
 */
class OdataResponseInBatch {
    /**
     * Creates an instance of OdataResponseInBatch.
     * @param {Object} inResponse response object
     * @param {string} responseId response ID
     */
    constructor(inResponse, responseId) {
        this._inResponse = inResponse;

        if (!responseId) throw new SerializationError('Invalid response ID');
        this._responseId = responseId;
    }

    /**
     * Returns the original incoming node response message.
     * @returns {Object} the original incoming response
     */
    getIncomingResponse() {
        return this._inResponse;
    }

    /**
     * Returns the response ID.
     * @returns {string} the response ID
     */
    getOdataResponseId() {
        return this._responseId;
    }

    /**
     * Returns the atomicity group to which this response belongs to.
     * @returns {string} the atomicity group ID
     */
    getAtomicityGroupId() {
        return this._atomicityGroupId;
    }

    /**
     * Sets the atomicity group to which this response belongs to.
     * @param {string} atomicityGroup atomicity group
     * @returns {OdataResponseInBatch} this instance
     * @package
     */
    setAtomicityGroupId(atomicityGroup) {
        this._atomicityGroupId = atomicityGroup;
        return this;
    }

    /**
     * Return the current status code of this response.
     * @returns {number} The current status code.
     */
    getStatusCode() {
        return this._inResponse.statusCode;
    }

    /**
     * Returns the status message of the HTTP response.
     * @returns {string} The status message
     */
    getStatusMessage() {
        return this._inResponse.statusMessage;
    }
}

module.exports = OdataResponseInBatch;
