// @sap/odata-server/lib/core/PlainHttpResponse.js
'use strict';

const Writable = require('stream').Writable;
const HttpStatusCode = require('@sap/odata-commons').http.HttpStatusCode;
const HttpStatusCodes = HttpStatusCode.StatusCodes;
const HttpStatusCodesText = HttpStatusCode.Texts;

const CRLF = '\r\n';
const HTTP_VERSION = 'HTTP/1.1';

/**
 * Simulates an http.ServerResponse; used for a part of a multipart/mixed response.
 * @extends Writable
 * @hideconstructor
 */
class PlainHttpResponse extends Writable {
    /**
     * Constructor
     */
    constructor() {
        super();
        this.statusCode = HttpStatusCodes.OK;  // as in node.js
        this._buffers = [];
        this._header = {};

        this.headersSent = false;  // as in node.js
    }

    /**
     * Writable implementation.
     * @param {Buffer|string} chunk the chunk to write
     * @param {string} encoding its encoding
     * @param {Function} callback the callback
     * @private
     */
    _write(chunk, encoding, callback) {
        this._buffers.push(chunk);
        callback();
    }

    /**
     * Returns the value of the requested header.
     * @param {string} name header name
     * @returns {string} header value
     */
    getHeader(name) {
        return this._header[name.toLowerCase()];
    }

    /**
     * Returns headers object.
     * @returns {Object} the headers
     */
    getHeaders() {
        return this._header;
    }

    /**
     * Sets a header.
     * @param {string} name the name of the header
     * @param {string} value the value of the header
     * @returns {PlainHttpResponse} this instance
     * @package
     */
    setHeader(name, value) {
        this._header[name.toLowerCase()] = value;
        return this;
    }

    /**
     * Returns the response body.
     * @returns {Buffer} the response body
     */
    getBody() {
        return Buffer.concat(this._buffers);
    }

    /**
     * Writes the response into a stream.
     * @param {Stream} stream the stream
     * @package
     */
    writeTo(stream) {
        // Write status line.
        const statusCode = this.statusCode || HttpStatusCodes.OK;
        const statusCodeText = HttpStatusCodesText[statusCode];

        stream.write(HTTP_VERSION + ' ' + statusCode + ' ' + statusCodeText + CRLF);

        // Write headers.
        // TODO: Check if content-length is a must.
        for (let name of Object.keys(this._header)) {
            if (name.toLowerCase() !== 'content-length') {
                stream.write(name + ': ' + this._header[name] + CRLF);
            }
        }

        stream.write(CRLF); // header body separator

        // Write payload.
        for (const buffer of this._buffers) stream.write(buffer);
    }
}

module.exports = PlainHttpResponse;
