// @sap/odata-server/lib/core/Service.js
'use strict';

const commons = require('@sap/odata-commons');
const validateThat = commons.validator.ParameterValidator.validateThat;
const RepresentationKinds = commons.format.RepresentationKind.Kinds;
const LoggerFacade = commons.logging.LoggerFacade;
const IllegalArgumentError = commons.errors.IllegalArgumentError;
const ComponentManager = require('./ComponentManager');
const MetadataCache = require('../utils/MetadataCache');
const PerformanceMonitorFactory = require('../utils/PerformanceMonitor').PerformanceMonitorFactory;
const ServiceResolutions = require('./ServiceResolutions');
const Context = require('./Context');
const CommandFactory = require('../invocation/CommandFactory');
const CommandExecutor = require('../invocation/CommandExecutor');
const FormatManager = require('../format/FormatManager');
const InternalServerError = require('../errors/InternalServerError');
const DebugLogger = require('../logging/DebugLogger');

const IS_DEBUG_REGEXP = new RegExp('[?&]odata-debug=(?:json|html)(?:&|$)');
const IS_PROFILING_REGEXP = new RegExp('[?&]odata-profile=true(?:&|$)');

/**
 * The definition of the Logger Object that can be registered via .log()
 *
 * @typedef {Object} Logger The logger object/class
 * @property {Function} logger.debug The debug function called with string
 * @property {Function} logger.path The path function called with string
 * @property {Function} logger.info The info function called with string
 * @property {Function} logger.warning The warning function called with string
 * @property {Function} logger.error The error function called with string
 * @property {Function} logger.fatal The fatal function called with string
 */

/**
 * The Service is the main service class where all service components must be registered.
 * This service is the main public API used by the service developer.
 */
class Service {
    /**
     * Constructor.
     * @param {EdmProvider} edmProvider the entity data model
     */
    constructor(edmProvider) {
        validateThat('edmProvider', edmProvider).truthy().typeOf('object');

        this._edmProvider = edmProvider;

        this._componentManager = new ComponentManager();

        this._formatManager = new FormatManager();
        this._bodyParserManager = new FormatManager();

        this._basePath = '';
        this._loggerFacade = null;

        this._metadataCache = null;
        this._metadataETag = null;

        this._isTrusted = false;
    }

    /**
     * Get the EDM.
     * @returns {Edm} the EDM
     */
    getEdm() {
        return this._edmProvider;
    }

    /**
     * Get the FormatManager.
     * @returns {FormatManager} the format manager
     */
    getFormatManager() {
        return this._formatManager;
    }

    /**
     * (Creates and) returns an instance of MetadataCache to cache MIME-type and locale specific metadata documents.
     * @returns {MetadataCache} the metadata cache
     * @private
     */
    _getMetadataCache() {
        if (!this._metadataCache) this._metadataCache = new MetadataCache();
        return this._metadataCache;
    }

    /**
     * Returns the base path of the current odata service.
     * The base path is the part of the URL not belonging to OData
     * which means that this part of the URL will not be interpreted as an OData resource.
     *
     * @returns {string} The base path of the url.
     */
    getBasePath() {
        return this._basePath;
    }

    /**
     * Sets the base path of the URL.
     * The OData resource URL starts after this base path and will be interpreted as an OData resource.
     * @param {string} basePath The base path of the url.
     * @returns {Service} This instance of odata service.
     * @throws {InternalServerError} if the provided basePath is not null and does not end with '/'
     */
    setBasePath(basePath) {
        if (basePath !== null && basePath !== undefined && !basePath.endsWith('/')) {
            throw new InternalServerError("Provided base path must end with '/'");
        }
        this._basePath = basePath;
        return this;
    }

    /**
     * Registers an event listener by its name.
     * The names are not arbitrary. Currently only the following events are supported:
     * General listeners: request, error, debug
     * CRUD listeners: create, read, update, delete
     * Batch listeners: batch-start, batch-end, atomicity-group-start, atomicity-group-end
     * @param {string} eventName The name of the listener.
     * @param {Object} eventListener the listener to register.
     * @returns {Service} Instance of this service.
     */
    on(eventName, eventListener) {
        const supportedEvents = [
            'request', 'error', 'debug', 'create', 'read', 'update', 'delete',
            'atomicity-group-start', 'atomicity-group-end', 'batch-start', 'batch-end'
        ];
        if (!supportedEvents.includes(eventName.toLowerCase())) {
            throw IllegalArgumentError.createForIllegalValue('eventName', ...supportedEvents);
        }
        validateThat('eventListener', eventListener).typeOf('function');

        this.getComponentManager().use(eventName, eventListener);
        this.getLogger().debug('Register listener for ' + eventName);
        return this;
    }

    /**
     * Registers a service component by its name. The names are not arbitrary. They must be used
     * from /lib/core/ComponentManager.Components.
     *
     * @param {string} name The name of the component.
     * @param {*} component the component to register.
     * @returns {Service} Instance of this service.
     */
    use(name, component) {
        this.getLogger().debug('Register component ' + name);
        this.getComponentManager().use(name, component);
        return this;
    }

    /**
     * Registers a logger. Occurring log events will be sent to this logger instance.
     * @param {Logger} logger The root logger
     * @param {Function} formatter Formatter to be used to convert
     * @returns {Service} Instance of this service.
     */
    log(logger, formatter) {
        this._loggerFacade = new LoggerFacade(logger).setFormatter(formatter);
        this._componentManager.use(ComponentManager.Components.LOGGER, this._loggerFacade);
        return this;
    }

    /**
     * Returns the component manager. The component manager manages all registered components within the library.
     * @returns {ComponentManager} The component manager
     */
    getComponentManager() {
        return this._componentManager;
    }

    /**
     * Sets the supported OData version.
     * @param {string} version - The supported OData version
     * @returns {Service} this instance of Service
     */
    setOdataVersion(version) {
        this._odataVersion = version;
        this.getLogger().debug('Set OdataVersion to ' + version);
        return this;
    }

    /**
     * Returns the supported OData version.
     * @returns {string} the supported OData version
     */
    getOdataVersion() {
        return this._odataVersion;
    }

    /**
     * Sets the distinct ETag for the metadata document.
     * @param {string} etag the ETag for the metadata document
     * @returns {Service} this instance of Service
     */
    setMetadataEtag(etag) {
        this._metadataETag = etag;
        return this;
    }

    /**
     * Returns the distinct ETag for the metadata document.
     * @returns {?string} the ETag for the metadata document
     */
    getMetadataEtag() {
        return this._metadataETag;
    }

    /**
     * Returns the current logger.
     * @returns {Logger} The current logger instance
     */
    getLogger() {
        return this._loggerFacade;
    }

    /**
     * Processes a request. This is the main starting point to use this library.
     * @param {http.IncomingMessage|PlainHttpRequest} request The request to process.
     * @param {http.ServerResponse|PlainHttpResponse} response The response to process.
     * @returns {Promise} a Promise that resolves when the request has been processed successfully
     */
    process(request, response) {
        const logger = this.getLogger();
        logger.path('Entering Service.process()...');

        // Check if odata-dabug=json|html is part of the request url
        const isDebugRequested = IS_DEBUG_REGEXP.test(request.url);
        const isProfiling = isDebugRequested || IS_PROFILING_REGEXP.test(request.url);

        logger.debug('Debug mode requested:', isDebugRequested);
        logger.debug('Profiling requested:', isProfiling);

        const rootPerformanceMonitor = PerformanceMonitorFactory.getInstance(isProfiling).start();

        logger.debug('Start processing request', request.method, request.url);
        logger.debug('Using OData base path', this.getBasePath());

        let context;
        let errorFromDebugHandler = null;
        let isDebug = false;

        return Promise.resolve()
            .then(() =>
                // This Promise handles the setup of the context object.
                new Promise(resolve => {
                    const serviceResolutionFn = ServiceResolutions.viaBasePath(this._basePath,
                        Boolean(request._batchContext));

                    context = new Context(request, response, this, serviceResolutionFn);
                    context.setPerformanceMonitor(rootPerformanceMonitor);
                    context.setLogger(isDebugRequested ? new DebugLogger(logger) : logger);
                    context.getRequest().setService(this);
                    logger.debug('Request headers:', context.getRequest().getHeaders());
                    resolve();
                }))
            .then(() =>
                // In this Promise, the debug handler gets called and the debug information gets evaluated.
                new Promise(resolve => {
                    const debugHandler = this.getComponentManager().getComponent('debug');
                    if (debugHandler) {
                        const next = error => {
                            if (error) {
                                logger.error(error.message);
                                logger.error('Error stacktrace:', ...error.stack.split('\n'));
                                errorFromDebugHandler = error;
                            } else {
                                // Check whether debug mode is allowed by the application (default is false) and
                                // if debug was requested. Only if both are true the request will be executed in debug mode
                                isDebug = context.getRequest().getContract().isDebug() && isDebugRequested;
                                context.getRequest().getContract().enableDebugMode(isDebug);
                                if (isDebug) {
                                    context.getResponse().setBuffered(true);
                                }
                                logger.path("Component 'debug' was called; debugging is "
                                    + (isDebug ? 'enabled' : 'disabled') + ' for the request');
                            }
                            resolve();
                        };

                        // Debug handler gets called here, with the next callback declared above
                        debugHandler(context.getRequest(), context.getResponse(), next);
                    } else {
                        resolve();
                    }
                }))
            .then(() => {
                // This Promise handles the creation and execution of the command chain.
                const options = {
                    context,
                    componentManager: this._componentManager,
                    formatManager: this._formatManager,
                    bodyParserManager: this._bodyParserManager,
                    isDebug: isDebug,
                    isProfiling
                };
                const successCommands = CommandFactory.createForSingleRequest(options);
                const errorCommands = CommandFactory.createForSingleRequestError(options);

                const executor = new CommandExecutor(context.getLogger(), rootPerformanceMonitor);
                return new Promise((resolve, reject) => {
                    const endCallback = err => {
                        rootPerformanceMonitor.stop();
                        if (err) {
                            logger.error(err.message);
                            logger.error('Error stacktrace:', ...err.stack.split('\n'));
                            reject(err);
                        } else {
                            logger.debug('Request processing finished');
                            resolve();
                        }
                    };
                    executor.execute(successCommands, errorCommands, errorFromDebugHandler, endCallback);
                });
            });
    }

    /**
     * Sets whether we trust in the data provided by the service implementation.
     * @param {boolean} isTrusted whether we trust in the data provided by the service implementation
     * @returns {Service} this instance
     */
    trust(isTrusted) {
        this._isTrusted = isTrusted;
        return this;
    }

    /**
     * Returns whether we trust in the data provided by the service implementation.
     * @returns {boolean} whether we trust in the data provided by the service implementation
     */
    isTrusted() {
        return this._isTrusted;
    }

    /**
     * Registers a serializer needed to serialize the result data into the response.
     * @param {RepresentationKind.Kinds} representationKind Kind of request like 'metadata' or 'resource'.
     * @param {string} contentType Requested content type like 'application/json'.
     * @param {Function} serializerFunction The function which will be called on serialization.
     * @param {Function} [parameterCheckFunction] the function to check the format parameters
     * @returns {Service} This instance of service.
     */
    format(representationKind, contentType, serializerFunction, parameterCheckFunction) {
        return this._registerDeserializer(this._formatManager,
            representationKind, contentType, serializerFunction, parameterCheckFunction);
    }

    /**
     * Registers a body parsing facade. This facade is responsible for parsing/deserializing the request payload.
     * @param {RepresentationKind.Kinds} representationKind The corresponding representation kind
     * @param {ContentTypeInfo.ContentTypes} contentType The corresponding content type
     * @param {Function} deserializerFunction The deserialization facade to register
     * @returns {Service} This instance
     */
    parse(representationKind, contentType, deserializerFunction) {
        return this._registerDeserializer(this._bodyParserManager,
            representationKind, contentType, deserializerFunction);
    }

    /**
     * Registers a provided deserializer function at its format manager in dependency of the corresponding
     * representation kind and content type
     *
     * @param {FormatManager} manager The corresponding manager
     * @param {RepresentationKind.Kinds} representationKind The corresponding representation kind
     * @param {ContentTypeInfo.ContentTypes} contentType The corresponding content type
     * @param {Function} parserFunction The parser facade to register
     * @param {?Function} parameterCheckFunction the function to check the format parameters
     * @returns {Service} This instance
     * @private
     */
    _registerDeserializer(manager, representationKind, contentType, parserFunction, parameterCheckFunction) {
        if (representationKind !== RepresentationKinds.NO_CONTENT) {
            validateThat('contentType', contentType).notNullNorUndefined().typeOf('string');
        }
        validateThat('serializerFunction', parserFunction).truthy().typeOf('function');

        manager.use(representationKind, contentType, parserFunction, parameterCheckFunction);

        return this;
    }
}

module.exports = Service;
