// @sap/cds-runtime/lib/cds-services/adapter/odata-v4/OData.js
const {
  BatchExitHandler: { ATOMICITY_GROUP_START, ATOMICITY_GROUP_END },
  Components: {
    DATA_CREATE_HANDLER,
    DATA_DELETE_HANDLER,
    DATA_READ_HANDLER,
    DATA_UPDATE_HANDLER,
    ACTION_EXECUTE_HANDLER,
    LOCALE_NEGOTIATOR,
    METADATA_HANDLER
  }
} = require('@sap/odata-server')
const request = require('./handlers/request')
const { end, start } = require('./handlers/atomicityGroup')

/**
 * Facade for creating an instance of a EDM based OData service.
 * @alias module:odata.OData
 */
class OData {
  /**
   * Constructs an OData service for the given EDM model.
   * @param {Object} edm - the EDM model.
   * @param {Object} [options] - optional object with options.
   * @param {Object} [options.logger] - optional logger object to be used in the odata library.
   * @param {string} [options.logLevel] - optional log level to be used according to winston/npm specification.
   * @param {string} [options.service] - Service name as specified in CSN.
   * @param {boolean} [options.crashOnError] - Application should crash on error. Defaults to true.
   *
   * @throws Error in case no or an invalid csn model is provided.
   */
  constructor (edm, csn, options = {}) {
    this._validateEdm(edm)
    this._options = options
    this._csn = csn
    this._createOdataService(edm)
  }

  _validateEdm (edm) {
    if (typeof edm !== 'object' || !edm.$Version) {
      const { getModelNotDefinedError } = require('../../util/errors')
      throw getModelNotDefinedError('EDM model')
    }
  }

  _createOdataService (edm) {
    const ServiceFactory = require('@sap/odata-server').ServiceFactory

    this._odataService = ServiceFactory.createService(
      edm,
      require('./utils/oDataConfiguration')(edm, this._csn, this._options)
    ).trust(process.env.NODE_ENV === 'production' || global.cds.env.odata.skipValidation === 'true')

    // will be added to express app like app.use('/base/path/', service) and odata-v4 wants app.use('/', service) if basePath is set
    this._odataService.setBasePath('/')
  }

  /**
   * The added cds service will be used at the handlers.
   * Some channel events have a 1:N relation to service handler events.
   * @param {Service} cdsService
   */
  addCDSServiceToChannel (cdsService) {
    this._odataService.log(cdsService.logger)
    this._cdsService = cdsService
    this._odataService.on('error', require('./handlers/error')(this._options.crashOnError))

    if (this._options.debug) {
      this._odataService.on('debug', require('./handlers/debug'))
    }

    this._odataService.on('request', request)
    this._odataService.on(ATOMICITY_GROUP_START, start)
    this._odataService.on(ATOMICITY_GROUP_END, end)

    this._odataService.use(LOCALE_NEGOTIATOR, require('./handlers/language')(this._options.defaultLocale))
    this._odataService.use(METADATA_HANDLER, require('./handlers/metadata')(cdsService, this._options))

    this._odataService.use(DATA_CREATE_HANDLER, require('./handlers/create')(cdsService, this._options))
    this._odataService.use(DATA_UPDATE_HANDLER, require('./handlers/update')(cdsService, this._options))
    this._odataService.use(DATA_DELETE_HANDLER, require('./handlers/delete')(cdsService, this._options))
    this._odataService.use(DATA_READ_HANDLER, require('./handlers/read')(cdsService, this._options))
    this._odataService.use(ACTION_EXECUTE_HANDLER, require('./handlers/action')(cdsService, this._options))
  }

  // _startPerfMeasurementOData (req) {
  //   if (req.performanceMeasurement) {
  //     const uuid = req.performanceMeasurement.uuid
  //     req.performanceMeasurement.performance.mark(`${uuid} ODataIn Start`)
  //   }
  // }

  /**
   * Process request.
   * @private
   */
  process (req, res) {
    // this._startPerfMeasurementOData(req)
    this._odataService.process(req, res).catch(err => {
      res.status(500)
      res.send({
        error: {
          code: 'null',
          message: err.message
        }
      })
    })
  }
}

module.exports = OData
