// @sap/cds-runtime/lib/cds-sql/sql-builder/BaseBuilder.js
const _slugify = name => name.replace(/::/g, '__').replace(/\./g, '_')

const quotingStyles = {
  quoted: (name, delimiter) => `${delimiter}${name}${delimiter}`,
  plain: name => _slugify(name),
  'all-upper': (name, delimiter) => `${delimiter}${_slugify(name.toUpperCase())}${delimiter}`,
  'all-lower': (name, delimiter) => `${delimiter}${_slugify(name.toLowerCase())}${delimiter}`,
  bracketed: name => `[${name}]`,
  hdbcds: (name, delimiter, csn) => {
    if (csn && csn.definitions && csn.definitions[name]) {
      return `${delimiter}${csn.definitions[name]['@cds.persistence.name']}${delimiter}`
    }
    return `${delimiter}${name}${delimiter}`
  }
}

/**
 * BaseBuilder class should not be instantiated. *
 */
class BaseBuilder {
  /**
   * The base class constructor for the builders.
   * If the options parameter is not specified, " are used as delimiter and ? as placeholders.
   *
   * @param {object} obj - The CQN object used for the insert operation
   * @param {object} [options] - The configuration object.
   * @param {string} [options.delimiter] - The delimiter string.
   * @param {string} [options.placeholder] - The placeholder for prepared statement.
   * @param {object} csn - The csn object
   */
  constructor (obj, options, csn) {
    this._obj = obj
    this._csn = csn
    const defaultOptions = {
      placeholder: '?',
      delimiter: '"',
      user: 'ANONYMOUS',
      now: { sql: 'NOW ()' }
    }
    this._options = Object.assign(defaultOptions, options)
    this._quotingStyle = global.cds.config.data.sql_mapping || 'plain'
    this._validateQuotingStyle()
  }

  _validateQuotingStyle () {
    let type = typeof this._quotingStyle
    if (type !== 'string' || !quotingStyles.hasOwnProperty(this._quotingStyle)) {
      type = type !== 'string' ? `Type ${type}` : `"${this._quotingStyle}"`
      throw new Error(`Quoting style: ${type} is not supported. Allowed strings: "quoted", "plain".`)
    }
  }

  _quoteElement (element) {
    return quotingStyles[this._quotingStyle](element, this._options.delimiter, this._csn)
  }
}

module.exports = BaseBuilder
