// @sap/cds/lib/srv/classes/index.js
const cds = require ('@sap/cds-reflect'), {lazified} = cds
const Service = require ('./Service')

require = lazified (module) // eslint-disable-line no-global-assign

module.exports = {
    Service,
    ServiceClient: require('./ServiceClient'),
    ServiceProvider: require('./ServiceProvider'),
    DatabaseService: require('./DatabaseService'),
    ApplicationService: require('./ApplicationService'),
    ProtocolAdapter: require('./ProtocolAdapter'),
}

// kind-specific registry for the service factory
cds.extend (Service.registry) .with (lazified ({
    [undefined]: require('./ApplicationService'),
    rest: require('./ServiceClient'),
    odata: require('./ServiceClient'),
    sqlite: require('./SQLDatabase'),
    hana: require('./SQLDatabase'),
    ['in-process-messaging']: require('./ServiceClient'), // FIXME: should go away
    ['file-based-messaging']: require('./ServiceClient'), // FIXME: should go away
    ['enterprise-messaging']: require('./ServiceClient'), // FIXME: should go away
}))
