// @sap/cds/lib/build/constants.js
exports.BUILD_OPTION_OUTPUT_MODE = "outputMode"

exports.OUTPUT_MODE_DEFAULT = "default"
exports.OUTPUT_MODE_PREVIEW = "preview"
exports.OUTPUT_MODE_RESULT_ONLY = "resultOnly"

exports.BUILD_TASK_USE_PREFIX = "@sap/cds/lib/build/"
exports.BUILD_TASK_JAVA = "java-cf"
exports.BUILD_TASK_NODE = "node-cf"
exports.BUILD_TASK_HANA = "hana"
exports.BUILD_TASK_FIORI = "fiori"
exports.BUILD_TASK_MTX = "mtx"

exports.ODATA_VERSION = "odata.version"
exports.ODATA_VERSION_V2 = "v2"
exports.ODATA_VERSION_V4 = "v4"

exports.BUILD_MODE_INPLACE = "inplace"
exports.BUILD_NODEJS_EDMX_GENERAION = "build.nodejs.edmxgeneration"

exports.FOLDER_GEN = "gen"

exports.FILE_EXT_CDS = ".cds"

exports.CDS_CONFIG_PATH_SEP = "/"

exports.SEVERITY_ERROR = "Error"
exports.SEVERITY_WARNING = "Warning"
exports.SEVERITY_INFO = "Info"
exports.SEVERITY_DEBUG = "Debug"
exports.SEVERITES = [exports.SEVERITY_ERROR, exports.SEVERITY_WARNING, exports.SEVERITY_INFO, exports.SEVERITY_DEBUG]
exports.LOG_LEVEL_ERROR = "error"
exports.LOG_LEVEL_WARN = "warn"
exports.LOG_LEVEL_INFO = "info"
exports.LOG_LEVEL_DEBUG = "debug"
exports.LOG_LEVELS = [exports.LOG_LEVEL_ERROR, exports.LOG_LEVEL_WARN, exports.LOG_LEVEL_INFO, exports.LOG_LEVEL_DEBUG]
