// @sap/cds-runtime/lib/cds-sql/expand/expandCQNToJoin.js
const getColumns = require('../utils/columns')
const { getOnCond } = require('../oncond/generateOnCond')
const sqliteLocales = ['de', 'fr']

const { ensureNoDraftsSuffix, ensureDraftsSuffix, ensureUnlocalized } = require('../utils/draftUtils')

// Symbols are used to add extra information in response structure
const GET_KEY_VALUE = Symbol.for('sap.cds.getKeyValue')
const TO_MANY = Symbol.for('sap.cds.toMany')

const SKIP_MAPPING = Symbol.for('sap.cds.skipMapping')
const IDENTIFIER = Symbol.for('sap.cds.identifier')
const IS_ACTIVE = Symbol.for('sap.cds.isActive')
const IS_UNION_DRAFT = Symbol.for('sap.cds.isUnionDraft')
const DRAFT_COLUMNS = ['IsActiveEntity', 'HasActiveEntity', 'HasDraftEntity', 'DraftAdministrativeData_DraftUUID']

class JoinCQNFromExpanded {
  constructor (cqn, csn, useWindow, locale) {
    this._useWindow = useWindow
    this._SELECT = Object.assign({}, cqn.SELECT)
    this._csn = cqn[Symbol.for('sap.cds.model')] || csn
    if (sqliteLocales.includes(locale)) {
      this._locale = locale
    }
    this.queries = []
    this.mappings = {}
  }

  /**
   * Build one to N queries and construct the post processing configs.
   * Each expand with a to many target will result in an extra query and config.
   */
  buildJoinQueries () {
    // side effect: this_aliases is set
    const aliases = this._getTableAlias(this._SELECT, [], this._getUnionTable(this._SELECT))

    // Add table aliases to all refs in where part obtained from annotations
    this._adaptAliasForWhere(this._SELECT.where)

    // Update elements at WHERE, so there are no issues with ambiguity
    this._adaptWhereOrderBy(this._SELECT, aliases)

    // Get first level of expanding regarding to many and all to one if not part of a nested to many expand.
    this._createJoinCQNFromExpanded(this._SELECT, [])

    return this
  }

  _getUnionTable (SELECT) {
    if (!SELECT.from.SET) {
      return
    }

    // Ensure the draft table is picked, no matter, which position
    for (const arg of SELECT.from.SET.args) {
      const { table } = this._getRef(arg.SELECT)
      // Do not handle non draft cases, as it will be unclear, which entity to pick from
      if (table.endsWith('_drafts')) {
        return table
      }
    }
  }

  _isDraftTargetActive (table) {
    return Boolean(this._csn.definitions[table])
  }

  _isDraftTree (table) {
    // TODO: this is a workaround until the service is flagged as draft enabled by cds-services
    if (!this._isDraft) {
      this._isDraft = this._isDraftEnabled(
        this._csn.definitions[ensureUnlocalized(table)] || this._csn.definitions[table.replace(/_drafts$/i, '')]
      )

      if (this._isDraft && !this._draftService) {
        this._draftService = table.replace(/\.\w+$/i, '')
      }
    }

    return this._isDraft
  }

  _isDraftEnabled (entity) {
    return Boolean(entity['@Common.DraftNode.PreparationAction'] || entity['@Common.DraftRoot.PreparationAction'])
  }

  /**
   * Build first level of expanding regarding to many and all to one if not part of a nested to many expand.
   * @param {Object} SELECT - SELECT part of a CQN.
   * @param {Array} toManyTree - Holds information how deeply nested the expand is and where the result is added in the tree.
   * @private
   */
  _createJoinCQNFromExpanded (SELECT, toManyTree) {
    const unionTable = this._getUnionTable(SELECT)
    const tableAlias = this._getTableAlias(SELECT, toManyTree, unionTable)
    const readToOneCQN = this._getReadToOneCQN(SELECT, tableAlias)
    const table = unionTable || this._getRef(SELECT).table
    const isDraftTree = this._isDraftTree(table)
    const entity = this._getEntityForTable(table)

    if (unionTable) {
      readToOneCQN[IS_UNION_DRAFT] = true
    }

    readToOneCQN[IS_ACTIVE] = isDraftTree ? this._isDraftTargetActive(table) : true

    const givenColumns = readToOneCQN.columns
    readToOneCQN.columns = []

    this._expandedToFlat({
      entity: entity,
      givenColumns: givenColumns,
      readToOneCQN: readToOneCQN,
      tableAlias: tableAlias,
      toManyTree: toManyTree
    })

    // Add at start, so that the deepest level is post processed first
    this.queries.push({
      SELECT: readToOneCQN,
      _toManyTree: toManyTree
    })
  }

  /**
   * Self referencing associations, two expanded entities based on same table, ...
   * Requires an abstract name to prevent ambiguity issues.
   * Use hash to prevent names longer than support by DB.
   * @returns {string}
   * @private
   */
  _getTableAlias (SELECT, toManyTree, unionTable) {
    return this._createAlias(toManyTree.length === 0 ? unionTable || this._getRef(SELECT).table : toManyTree.join(':'))
  }

  _getRef (SELECT) {
    const table = SELECT.from.hasOwnProperty('join') ? this._getRefFromJoin(SELECT.from.args) : SELECT.from

    return {
      table: table.SELECT ? this._getRef(table.SELECT).table : table.ref[0],
      as: table.as
    }
  }

  _getRefFromJoin (args) {
    if (args[0].join) {
      return this._getRefFromJoin(args[0].args)
    }

    if (args[0].ref) {
      return args[0]
    }

    // Order is reversed
    return args[args.length - 1]
  }

  /**
   * Create an alias from value.
   * @returns {string}
   * @private
   */
  _createAlias (value) {
    if (!this._aliases) {
      this._aliases = {}
    }

    if (!this._aliases[value]) {
      this._aliases[value] = String.fromCharCode(Object.keys(this._aliases).length + 97)
    }

    return this._aliases[value]
  }

  _getEntityForTable (table) {
    if (table === 'DraftAdministrativeData') {
      table = `DRAFT.${table}`
    }

    if (this._isDraft) {
      return this._csn.definitions[table] || this._csn.definitions[table.replace(/_drafts/i, '')]
    }

    return this._csn.definitions[table]
  }

  /**
   * Get base CQN, with the same filters as origin.
   * @param {Object} SELECT
   * @param {string} tableAlias
   * @returns {Object}
   * @private
   */
  _getReadToOneCQN (SELECT, tableAlias) {
    const cqn = Object.assign({}, SELECT, { from: Object.assign({}, SELECT.from) })

    if (cqn.from.hasOwnProperty('join')) {
      this._adaptJoin(tableAlias, cqn)
    } else {
      if (cqn.from.SET) {
        cqn.from.SET = Object.assign({}, cqn.from.SET, { args: this._adaptUnionArgs(cqn.from.SET.args) })
      }

      cqn.from.as = tableAlias
    }

    return cqn
  }

  _adaptTableNameInColumn (column, originalIdentifier, tableAlias) {
    return column.ref && column.ref[0] === originalIdentifier
      ? Object.assign({}, column, { ref: [tableAlias, column.ref[1]] })
      : column
  }

  _adaptJoin (tableAlias, cqn) {
    cqn.from.args = cqn.from.args.slice(0)

    const index = cqn.from.args[0].ref ? 0 : cqn.from.args.length - 1
    const target = Object.assign({}, cqn.from.args[index], { as: tableAlias })
    const originalIdentifier = cqn.from.args[index].as || cqn.from.args[index].ref[0]

    cqn.from.args[index] = target
    cqn.from.on = cqn.from.on.map(column => this._adaptTableNameInColumn(column, originalIdentifier, tableAlias))
    cqn.columns = cqn.columns.map(column => this._adaptTableNameInColumn(column, originalIdentifier, tableAlias))
  }

  _adaptUnionArgs (args) {
    return args.map(arg => {
      if (arg.SELECT.columns) {
        // remove the expands from the sub selects, as they are joined against the unioned result
        arg = Object.assign({}, arg, { SELECT: Object.assign({}, arg.SELECT) })
        arg.SELECT.columns = arg.SELECT.columns.filter(element => {
          return !element.expand || typeof element.expand === 'function'
        })
      }

      return arg
    })
  }

  /**
   * Ensure that columns are accesed in combination with table alias.
   * Prevents ambiquity issues.
   * @param {Object} cqn
   * @param {string} tableAlias
   * @returns {Object}
   * @private
   */
  _adaptWhereOrderBy (cqn, tableAlias) {
    if (cqn.where) {
      cqn.where = cqn.where.map(element => {
        if (element.list) {
          return Object.assign(element, {
            list: element.list.map(element => this._checkOrderByWhereElementRecursive(cqn, element, tableAlias))
          })
        }
        return this._checkOrderByWhereElementRecursive(cqn, element, tableAlias)
      })
    }

    if (cqn.orderBy) {
      cqn.orderBy = cqn.orderBy.map(element => {
        return this._checkOrderByWhereElementRecursive(cqn, element, tableAlias)
      })
    }

    return cqn
  }

  _addAliasAndDeleteSymbol (whereElement) {
    this._addAlias(whereElement)
    delete whereElement.ref[Symbol.for('sap.cds.FROM_ANNOTATION')]
  }

  _addAlias (whereElement) {
    whereElement.ref && whereElement.ref.splice(0, 1, Object.values(this._aliases)[0])
  }

  _adaptAliasForFrom (from) {
    if (from.args) {
      from.args.forEach(arg => {
        this._adaptAliasForFrom(arg)
      })
    } else if (from.SELECT) {
      this._adaptAliasForFrom(from.SELECT.from)
      if (from.SELECT.where) {
        this._adaptAliasForWhere(from.SELECT.where)
      }
    }
  }

  _adaptAliasForWhere (where) {
    if (where) {
      for (const whereElement of where) {
        if (whereElement.ref && whereElement.ref[Symbol.for('sap.cds.FROM_ANNOTATION')] === true) {
          this._addAliasAndDeleteSymbol(whereElement)
        }

        if (whereElement.SELECT) {
          if (whereElement.SELECT.where) {
            this._adaptAliasForWhere(whereElement.SELECT.where)
          }
          this._adaptAliasForFrom(whereElement.SELECT.from)
        }
      }
    }
  }

  _navigationNeedsAlias (element, { table } = {}) {
    const entity = this._csn.definitions[table]
    if (entity) {
      const e = this._csn.definitions[table].elements[element.ref[0]]
      return e && (e.type === 'cds.Association' || e.type === 'cds.Composition')
    }

    return false
  }

  _checkOrderByWhereElementRecursive (cqn, element, tableAlias) {
    if (element.func) {
      element = Object.assign({}, element)
      this._functionNeedsReplacement(cqn, tableAlias, element)
    } else if (element.ref) {
      element = Object.assign({}, element)
      element.ref = element.ref.slice(0)

      if (element.ref.length === 1) {
        element.ref.unshift(tableAlias)
      } else if (this._elementAliasNeedsReplacement(element, this._getUnionTable(cqn) || this._getRef(cqn))) {
        element.ref[0] = tableAlias
      } else if (this._navigationNeedsAlias(element, this._getUnionTable(cqn) || this._getRef(cqn))) {
        element.ref.unshift(tableAlias)
      }

      this._functionNeedsReplacement(cqn, tableAlias, element)
    } else if (element.xpr) {
      element = Object.assign({}, element)
      element.xpr = element.xpr.map(nestedElement => {
        return this._checkOrderByWhereElementRecursive(cqn, nestedElement, tableAlias)
      })
    } else if (element.SELECT && element.SELECT.where) {
      element = {
        SELECT: Object.assign({}, element.SELECT, {
          where: this._adaptWhereSELECT(this._getUnionTable(cqn) || this._getRef(cqn), element.SELECT.where, tableAlias)
        })
      }
    }

    return element
  }

  /**
   * Change alias of most outer table query to md5 sum.
   * @private
   */
  _adaptWhereSELECT (aliasedTable, where, tableAlias) {
    return where.map(element => {
      return this._elementAliasNeedsReplacement(element, aliasedTable)
        ? Object.assign({}, element, { ref: [tableAlias, element.ref[1]] })
        : element
    })
  }

  _elementAliasNeedsReplacement (element, { table, as }) {
    // ref contains a single column, no replacement needed
    if (!element.ref || element.ref.length < 2) {
      return false
    }

    switch (element.ref[0]) {
      case table:
      case as:
        return true
      default:
        return false
    }
  }

  _isValidFunc (element) {
    if (typeof element.func === 'string' && Array.isArray(element.args)) {
      return true
    }

    if (
      typeof element.ref[0] === 'string' &&
      typeof element.ref[1] === 'object' &&
      Array.isArray(element.ref[1].args)
    ) {
      return true
    }
  }

  _mapArg (arg, cqn, tableAlias) {
    if (Array.isArray(arg.list)) {
      arg = Object.assign({}, arg)
      arg.list = arg.list.map(item => {
        return this._checkOrderByWhereElementRecursive(cqn, item, tableAlias)
      })

      return arg
    }

    return this._checkOrderByWhereElementRecursive(cqn, arg, tableAlias)
  }

  _functionNeedsReplacement (cqn, tableAlias, element) {
    if (!this._isValidFunc(element)) {
      return
    }

    if (element.ref) {
      element.ref[1] = Object.assign({}, element.ref[1])
      element.ref[1].args = element.ref[1].args.map(arg => {
        return this._mapArg(arg, cqn, tableAlias)
      })
    } else {
      element.args = element.args.slice(0)
      element.args = element.args.map(arg => {
        return this._mapArg(arg, cqn, tableAlias)
      })
    }
  }

  _skip (targetEntity) {
    return targetEntity && targetEntity['@cds.persistence.skip'] === true
  }

  /**
   * Build CQN(s) with JOINs for expanding. In case of expanding with to many an additional CQN will be pushed to toManyCQN.
   * @param {Object} arg - Avoiding many arguments and issues that come with it by using an object.
   * @param {Object} arg.entity - Entity that is taken from CSN.
   * @param {Array} arg.givenColumns - List of read columns taken from CQN.
   * @param {Object} arg.readToOneCQN - Build CQN the JOIN(s) should be added to or it will be used to filter an expanded to many entity.
   * @param {string} arg.tableAlias - Table alias
   * @param {Array} arg.toManyTree - Information, where the expand array is located in the result array.
   * @returns {Object}
   * @private
   */
  _expandedToFlat ({ entity, givenColumns, readToOneCQN, tableAlias, toManyTree }) {
    const toManyColumns = []
    const mappings = this._getMappingObject(toManyTree)

    for (const column of givenColumns) {
      if (column.expand) {
        const target = entity.elements[column.ref[0]] ? entity.elements[column.ref[0]].target : column.ref[0]
        const targetEntity = this._csn.definitions[target]
        if (this._skip(targetEntity)) continue
      }

      // To many can only be build, once all other columns have been processed.
      if (this._isExpandToMany(column, entity)) {
        mappings[column.ref[0]] = { [TO_MANY]: true }
        toManyColumns.push({ parentAlias: tableAlias, column: column })

        // Expands with to one target can be processed directly
      } else if (typeof column.expand === 'object') {
        this._addJoinAndElements({
          column,
          entity,
          readToOneCQN,
          toManyTree,
          parentAlias: tableAlias
        })
      } else {
        // No expand, directly add the column and its mapping.
        readToOneCQN.columns.push(this._addAliasToColumn(column, entity, tableAlias, mappings))
      }
    }

    // only as second step handle expand to many, or else keys might still be unknown
    this._toMany({ entity, readToOneCQN, tableAlias, toManyColumns, toManyTree, mappings })
  }

  /**
   * Follow the tree to get to the relevant config object.
   * @param {Array} toManyTree
   * @returns {Object}
   * @private
   */
  _getMappingObject (toManyTree) {
    let mappings = this.mappings

    for (const element of toManyTree) {
      if (!mappings[element]) {
        mappings[element] = {}
      }

      mappings = mappings[element]
    }

    return mappings
  }

  _addJoinCompToOne (cqn, entity, tableAlias) {
    const draftTable = ensureDraftsSuffix(entity.target)
    const on = []
    for (const key of Object.keys(entity._target.keys)) {
      if (key !== 'IsActiveEntity') {
        on.push({ ref: ['drafts', key] }, '=', { ref: [tableAlias, key] })
      }
    }
    return {
      args: [cqn, { ref: [draftTable], as: 'drafts' }],
      join: 'left',
      on: on
    }
  }

  _isExpandToMany (column, entity) {
    return typeof column.expand === 'function' || (column.expand && column.ref[0] === 'DraftAdministrativeData')
      ? false
      : column.expand && entity.elements[column.ref[0]].is2many
  }

  _isNavigationToOne (activeTable, target) {
    return (
      activeTable && target && (target.type === 'cds.Composition' || target.type === 'cds.Association') && target.is2one
    )
  }

  /**
   * Adds JOIN instructions to CQN for expands with 1:1 target and returns config how to map it back.
   * @returns {Object}
   * @private
   */
  _addJoinAndElements ({ column, entity, readToOneCQN, toManyTree, parentAlias }) {
    const extendedToManyTree = toManyTree.concat(column.ref)
    const tableAlias = this._createAlias(extendedToManyTree.join(':'))
    const target = entity.elements[column.ref[0]] ? entity.elements[column.ref[0]].target : column.ref[0]

    // if union always only expand with active, otherwise evaluate flag
    // if flag shows false, we check entity for associations to non draft
    const activeTableRequired =
      readToOneCQN[IS_UNION_DRAFT] || readToOneCQN[IS_ACTIVE] || !this._isDraftEnabled(this._csn.definitions[target])

    // TODO: If draft union and composition target add union as to be joined
    readToOneCQN.from = {
      args: [
        readToOneCQN.from.SET ? this._unionToSubQuery(readToOneCQN) : readToOneCQN.from,
        { ref: [this._refFromRefByExpand(column.ref[0], entity.elements, activeTableRequired)], as: tableAlias }
      ],
      join:
        column.ref[0] === 'DraftAdministrativeData' || !entity.elements[column.ref[0]].notNull || this._isDraft
          ? 'left'
          : 'inner',
      on: this._getOnCond(entity, column.ref[0], tableAlias, parentAlias, readToOneCQN)
    }

    if (column.ref[0] !== 'DraftAdministrativeData') {
      this._addJoinKeyColumnsToUnion(readToOneCQN.from.args, readToOneCQN.from.on, parentAlias)
    }

    // special case of navigation to one requires additional LEFT JOIN and CASE for HasDraftEntity
    const compToOne = this._isNavigationToOne(readToOneCQN[IS_ACTIVE], entity.elements[column.ref[0]])
    const index = column.expand.findIndex(col => col.ref[0] === 'HasDraftEntity')

    if (compToOne && index !== -1) {
      readToOneCQN.from = this._addJoinCompToOne(readToOneCQN.from, entity.elements[column.ref[0]], tableAlias)
      column.expand[index] = {
        xpr: [
          'case',
          'when',
          'drafts.DraftAdministrativeData_DraftUUID',
          'IS NOT NULL',
          'then',
          'true',
          'else',
          'false',
          'end'
        ],
        as: 'HasDraftEntity',
        cast: { type: 'cds.Boolean' }
      }
    }

    this._expandedToFlat({
      entity: this._getEntityForTable(target),
      givenColumns: column.expand.map(col => {
        if (
          activeTableRequired &&
          col.ref &&
          col.ref.length &&
          (col.ref[0] === 'IsActiveEntity' || col.ref[0] === 'HasActiveEntity')
        ) {
          return {
            val: col.ref[0] === 'IsActiveEntity',
            as: col.ref[0],
            cast: { type: 'cds.Boolean' }
          }
        }
        return col
      }),
      readToOneCQN: readToOneCQN,
      tableAlias: tableAlias,
      toManyTree: extendedToManyTree
    })
  }

  _refFromRefByExpand (column, elements, isActiveRequired = true) {
    if (column === 'DraftAdministrativeData') {
      return 'DRAFT.DraftAdministrativeData'
    }

    let target = elements[column].target

    if (isActiveRequired && this._csn.definitions[target]['@cds.localized'] !== false) {
      const locale = this._locale ? `${this._locale}.` : ''
      const localized = `localized.${locale}${target}`
      if (this._csn.definitions[localized]) {
        target = localized
      }
    }

    return `${target}${isActiveRequired ? '' : '_drafts'}`
  }

  _unionToSubQuery (readToOneCQN) {
    return {
      SELECT: {
        columns: Array.from(readToOneCQN.columns),
        from: readToOneCQN.from,
        as: readToOneCQN.from.as
      }
    }
  }

  _getAliases (columns) {
    return columns.reduce((aliases, entry) => {
      if (!entry.ref) {
        return aliases
      }

      if (!aliases[entry.ref[0]]) {
        aliases[entry.ref[0]] = {}
      }

      aliases[entry.ref[0]][entry[IDENTIFIER]] = entry.as
      return aliases
    }, {})
  }

  _getSubSelectColumns (cqn) {
    const args = cqn.args || cqn.from.args

    if (args) {
      for (const arg of args) {
        if (arg.ref) {
          continue
        }

        if (arg.SELECT && arg.SELECT.columns.some(column => column[IDENTIFIER])) {
          return arg.SELECT.columns
        }

        return this._getSubSelectColumns(arg.SELECT || arg)
      }
    }

    if (!cqn.from) return []

    const columns = cqn.from.SELECT ? cqn.from.SELECT.columns : cqn.columns
    return columns.some(column => column[IDENTIFIER]) ? columns : []
  }

  _getOnCond (entity, column, tableAlias, parentAlias, readToOneCQN) {
    if (column === 'DraftAdministrativeData') {
      if (readToOneCQN[IS_ACTIVE]) {
        const subWhere = []

        for (const key of Object.keys(entity.keys)) {
          if (key !== 'IsActiveEntity') {
            subWhere.push({ ref: [key] }, '=', { ref: [parentAlias, key] })
          }
        }

        return [
          { ref: [tableAlias, 'DraftUUID'] },
          'in',
          {
            SELECT: {
              from: { ref: [ensureDraftsSuffix(entity.name)] },
              columns: [{ ref: ['DraftAdministrativeData_DraftUUID'] }],
              where: subWhere
            }
          }
        ]
      }

      return [{ ref: [tableAlias, 'DraftUUID'] }, '=', { ref: [parentAlias, 'DraftAdministrativeData_DraftUUID'] }]
    }

    const elements = entity.elements

    // No window function/ sub select
    const subSelectColumns = this._getSubSelectColumns(readToOneCQN)

    if (subSelectColumns.length === 0) {
      return getOnCond(elements[column], column, this._csn, tableAlias, parentAlias)
    }

    const aliases = this._getAliases(subSelectColumns)

    const on = getOnCond(elements[column], column, this._csn, tableAlias, parentAlias)

    for (const element of on) {
      if (element.ref && aliases[element.ref[0]] && aliases[element.ref[0]][element.ref[1]]) {
        element.ref[1] = aliases[element.ref[0]][element.ref[1]]
      }
    }

    return on
  }

  _addJoinKeyColumnsToUnion (args, on, parentAlias) {
    for (const arg of args) {
      if (arg.ref) {
        continue
      }

      if (arg.args) {
        this._addJoinKeyColumnsToUnion(arg.args, on, parentAlias)
      } else if (arg.SELECT.from.SET && arg.SELECT.as === parentAlias) {
        this._addColumns(arg.SELECT.from.SET.args, on, parentAlias)
      }
    }
  }

  _addColumns (args, on, parentAlias) {
    const [
      {
        SELECT: { columns }
      }
    ] = args
    const keyColumns = on
      .filter(entry => {
        return (
          entry.ref &&
          entry.ref[0] === parentAlias &&
          !columns.some(column => column.ref && column.ref[column.ref.length - 1] === entry.ref[1])
        )
      })
      .map(entry => ({ ref: [entry.ref[1]] }))

    if (keyColumns.length === 0) return

    for (const {
      SELECT: { columns }
    } of args) {
      columns.push(...keyColumns)
    }
  }

  /**
   * Add an unique alias to each column, to avoid ambiguity.
   * Add this information to the post process config.
   * @returns {Object}
   * @private
   */
  _addAliasToColumn (column, entity, tableAlias, mappings) {
    // No identifier for this row entry or technical column
    if (this._isAliasNotNeeded(column)) {
      return column
    }

    return this._buildNewAliasColumn(column, entity, tableAlias, mappings)
  }

  /**
   * Technical or a value without a casted ensureNoDraftsSuffixname, or some other not yet supported combinations should not be refactored.
   * @param {Object} column
   * @returns {boolean}
   * @private
   */
  _isAliasNotNeeded (column) {
    // functions, direct values, ...
    if (!column.ref && !column.as) {
      return true
    }

    // No column name specified means false
    return column.ref && typeof column.ref[column.ref.length - 1] !== 'string'
  }

  _buildNewAliasColumn (column, entity, tableAlias, mappings) {
    // Casted name, vs column name
    const identifier = this._getIdentifier(column, tableAlias)
    const as = column.as && !DRAFT_COLUMNS.includes(column.as) ? column.as : `${tableAlias}_${identifier}`
    const aliasedElement = Object.assign({}, column)
    aliasedElement.as = as

    // Add table alias or name to handle cases, where joined tables have same column names
    if (this._isElement(column.ref, entity)) {
      const alias = tableAlias || ensureNoDraftsSuffix(entity.name)
      aliasedElement.ref = alias ? [alias, column.ref[0]] : [column.ref[0]]
    }

    if (!column[SKIP_MAPPING]) {
      mappings[column[IDENTIFIER] || identifier] = as
    }

    return aliasedElement
  }

  _getIdentifier (column, tableAlias) {
    if (column.as) {
      return column.as.startsWith(`${tableAlias}_`) ? column.ref[column.ref.length - 1] : column.as
    }

    return column.ref[column.ref.length - 1]
  }

  _isElement (ref, entity) {
    if (!ref || ref.length !== 1) {
      return false
    }

    // Normal element
    if (entity.elements[ref[0]]) {
      return true
    }

    // Draft column
    return DRAFT_COLUMNS.includes(ref[0])
  }

  _getKeyNames (entity) {
    const keys = entity.keys

    if (!keys) {
      return
    }

    const keyNames = []

    for (const key of Object.keys(keys)) {
      if (keys[key].foreignKeys) {
        for (const foreign of Object.keys(keys[key].foreignKeys)) {
          // OLD CSN
          keyNames.push(`${key}_${foreign}`)
        }
      } else if (!keys[key].keys) {
        keyNames.push(key)
      }
    }

    return keyNames
  }

  _toMany ({ entity, readToOneCQN, tableAlias, toManyColumns, toManyTree, mappings }) {
    if (toManyColumns.length === 0) {
      return
    }

    this._addKeysIfNeeded({ entity, readToOneCQN, tableAlias })

    for (const { column, parentAlias } of toManyColumns) {
      this._createJoinCQNFromExpanded(
        this._buildExpandedCQN({
          column,
          entity,
          readToOneCQN,
          toManyTree,
          mappings,
          parentAlias
        }),
        toManyTree.concat([column.ref[0]])
      )
    }
  }

  /**
   * In case of to many relations, a key is needed for post processing.
   * @private
   */
  _addKeysIfNeeded ({ entity, readToOneCQN, tableAlias }) {
    for (const name of this._getMissingKeys({ entity, readToOneCQN, tableAlias })) {
      if (name === 'IsActiveEntity') {
        readToOneCQN.columns.push({
          val: readToOneCQN[IS_ACTIVE],
          as: 'IsActiveEntity',
          cast: { type: 'cds.Boolean' }
        })
      } else {
        readToOneCQN.columns.push({
          as: `${tableAlias}_${name}`,
          ref: [tableAlias, name]
        })
      }
    }
  }

  /**
   * Compare the list of available keys with keys listed already listed at CQN and return missing.
   * @param entity
   * @returns {Array}
   * @private
   */
  _getMissingKeys ({ entity, readToOneCQN, tableAlias }) {
    const keyNames = this._getKeyNames(entity)

    if (!keyNames) {
      return
    }

    return keyNames.filter(name => {
      return !readToOneCQN.columns.some(column => column.as === `${tableAlias}_${name}` || column.as === name)
    })
  }

  /**
   * Construct the base CQN for a to many expands.
   * @returns {Object}
   * @private
   */
  _buildExpandedCQN ({ column, entity, readToOneCQN, toManyTree, mappings, parentAlias }) {
    const ref = this._getJoinRef(entity.elements, column.ref[0], readToOneCQN[IS_ACTIVE])
    const tableAlias = this._createAlias(toManyTree.concat(column.ref).join(':'))
    const on = getOnCond(entity.elements[column.ref[0]], column.ref[0], this._csn, tableAlias, 'filterExpand')
    const filterExpand = this._getFilterExpandCQN(readToOneCQN, on, parentAlias, entity.keys)
    const expandedEntity = this._csn.definitions[entity.elements[column.ref[0]].target]
    const joinColumns = this._getJoinColumnsFromOnAddToMapping(mappings[column.ref[0]], parentAlias, on, entity)

    const cqn = {
      from: {
        join: 'inner',
        args: [{ ref: [ref], as: tableAlias }, filterExpand],
        on: on
      }
    }

    if (typeof readToOneCQN[IS_ACTIVE] === 'boolean') {
      cqn[IS_ACTIVE] = !ref.endsWith('_drafts')
    }

    cqn.columns = this._getColumnsForExpand({
      tableAlias,
      columnList: column,
      entity: expandedEntity,
      joinColumns,
      isActive: cqn[IS_ACTIVE],
      parentEntity: entity
    })

    if (column.where) {
      cqn.where = this._copyWhere(column.where)
    }

    if (column.orderBy) {
      cqn.orderBy = this._copyOrderBy(column.orderBy, tableAlias, expandedEntity)
    }

    if (column.limit && !this._useWindow) {
      this._addLimitToCqn(cqn, column, tableAlias, expandedEntity)
    }

    return this._adaptWhereOrderBy(this._addWindowIfNeeded(cqn, column, tableAlias, expandedEntity), tableAlias)
  }

  _getJoinRef (elements, column, isActive) {
    if (typeof isActive !== 'boolean' || isActive || elements[column].type !== 'cds.Composition') {
      return elements[column].target
    }

    return `${this._draftService}.${elements[column].target.split('.').pop()}_drafts`
  }

  /**
   * Solve limit by group with a sub select at the join clause.
   * @private
   */
  _addLimitToCqn (cqn, column, tableAlias, expandedEntity) {
    const columns = this._getKeyColumnForTarget(tableAlias, cqn[IS_ACTIVE], expandedEntity)
    const inSelect = this._getLimitInSelect(cqn, columns, column.limit, column.orderBy, expandedEntity)

    cqn.from.on.push('and')
    cqn.from.on.push({
      list: columns
    })
    cqn.from.on.push('IN', inSelect)
  }

  /**
   * Get the list of key columns in ref format.
   * Add the table alias to avoid ambiquity issues.
   * @return {Array}
   * @private
   */
  _getKeyColumnForTarget (tableAlias, isActive, expandedEntity) {
    return this._getKeyNames(expandedEntity)
      .filter(column => typeof isActive !== 'boolean' || column !== 'IsActiveEntity')
      .map(column => {
        return { ref: [tableAlias, column] }
      })
  }

  _getLimitInSelect (cqn, columns, limit, orderBy, expandedEntity) {
    const select = {
      SELECT: {
        columns: this._copyColumns(columns, 'limitFilter'),
        from: { ref: [cqn.from.args[0].ref[0]], as: 'limitFilter' },
        where: this._convertOnToWhere(cqn.from.on, cqn.from.args[0].as, 'limitFilter'),
        limit: limit
      }
    }

    if (orderBy) {
      select.SELECT.orderBy = this._copyOrderBy(orderBy, 'limitFilter', expandedEntity)
    }

    return select
  }

  _isPathExpressionToOne (ref, entity) {
    const ref0 = ref[0]
    const el = entity.elements[ref0]
    return el && el.is2one
  }

  _copyOrderBy (orderBy, alias, expandedEntity) {
    return orderBy.map(element => {
      const ref =
        element.ref.length === 1
          ? [alias, element.ref[0]]
          : this._isPathExpressionToOne(element.ref, expandedEntity)
            ? [alias, ...element.ref]
            : [alias, element.ref[1]]
      return { ref, sort: element.sort }
    })
  }

  _getWindowRef (tableAlias, name) {
    return { [IDENTIFIER]: name, ref: [tableAlias, name], as: `${tableAlias}_${name}` }
  }

  _getWindowVal (tableAlias, name, val) {
    return { [IDENTIFIER]: name, val: val, as: `${tableAlias}_${name}` }
  }

  _draftColumnsToCQN (cqn, tableAlias, expandedEntity) {
    return DRAFT_COLUMNS.map(name => {
      if (name === 'IsActiveEntity') {
        return this._getWindowVal(tableAlias, name, cqn[IS_ACTIVE])
      }

      if (cqn[IS_ACTIVE]) {
        if (name === 'DraftAdministrativeData_DraftUUID') {
          return this._getWindowVal(tableAlias, name, null)
        }
        if (name === 'HasActiveEntity') {
          return this._getWindowVal(tableAlias, name, false)
        }
        if (name === 'HasDraftEntity') {
          return this._getHasDraftEntityXpr(expandedEntity, tableAlias)
        }
      }

      return this._getWindowRef(tableAlias, name)
    })
  }

  _getHasDraftEntityXpr (expandedEntity, tableAlias, columns = []) {
    const draftTable = ensureDraftsSuffix(expandedEntity.name)
    const where = Object.keys(expandedEntity.keys)
      .filter(name => name !== 'IsActiveEntity')
      .reduce((res, keyName) => {
        let keyAlias
        if (this._useWindow) {
          const element = columns.find(
            col => col.as && col.ref && col.ref.length === 2 && col.ref[1] === keyName && col.ref[0] === tableAlias
          )
          keyAlias = element && element.as
        }
        res.push({ ref: [draftTable, keyName] }, '=', { ref: [tableAlias, keyAlias || keyName] })
        return res
      }, [])
    const hasDraftQuery = {
      SELECT: {
        from: { ref: [draftTable] },
        columns: [{ val: 1 }],
        where: where
      }
    }
    return {
      xpr: ['case', 'when', hasDraftQuery, 'IS NOT NULL', 'then', 'true', 'else', 'false', 'end'],
      as: 'HasDraftEntity',
      cast: { type: 'cds.Boolean' }
    }
  }

  _getWindowPartitions (on) {
    const partitions = []

    for (const element of on) {
      if (
        element.ref &&
        element.ref[0] !== 'filterExpand' &&
        element.ref[element.ref.length - 1] !== 'IsActiveEntity'
      ) {
        if (partitions.length !== 0) {
          partitions.push(',')
        }

        partitions.push(element)
      }
    }

    return partitions
  }

  _getWindowColumns (cqn, column, tableAlias, expandedEntity) {
    const columns = cqn.columns
      .filter(entry => entry.val || (entry.ref && !entry.expand && entry.ref[0] !== tableAlias))
      .concat(
        getColumns(expandedEntity).map(element => {
          return this._getWindowRef(tableAlias, element.name)
        })
      )

    if (this._isDraftEnabled(expandedEntity)) {
      columns.push(...this._draftColumnsToCQN(cqn, tableAlias, expandedEntity))
    }

    columns.push(
      this._getWindowXpr(
        this._getWindowPartitions(cqn.from.on),
        this._copyOrderBy(column.orderBy || [], tableAlias, expandedEntity)
      )
    )

    return columns
  }

  _getWhereForLimit (column) {
    const offset = (column.limit.offset && column.limit.offset.val) || 0
    return [
      { ref: ['rowNumber'] },
      '>',
      { val: offset },
      'and',
      { ref: ['rowNumber'] },
      '<',
      { val: offset + column.limit.rows.val + 1 }
    ]
  }

  _addWindowIfNeeded (cqn, column, tableAlias, expandedEntity) {
    if (!this._useWindow || (!column.orderBy && !column.limit)) {
      return cqn
    }

    const windowCQN = {
      columns: cqn.columns.map(entry => {
        if (entry.ref && entry.as) {
          const newRef = entry.ref[0] === 'filterExpand' ? [entry.as] : [entry.ref[0], entry.as]
          return Object.assign({}, entry, { [IDENTIFIER]: entry.ref[1], ref: newRef })
        }

        if (entry.val) {
          const returnValue = { [IDENTIFIER]: entry.as, ref: [tableAlias, entry.as], as: entry.as }
          if (SKIP_MAPPING in entry) {
            returnValue[SKIP_MAPPING] = entry[SKIP_MAPPING]
          }
          return returnValue
        }

        return Object.assign({}, entry)
      }),
      from: {
        SELECT: cqn
      }
    }

    cqn.columns = this._getWindowColumns(cqn, column, tableAlias, expandedEntity)

    if (column.limit) {
      windowCQN.where = this._getWhereForLimit(column)
    }

    return windowCQN
  }

  _getOrderByForWindowFn (orderBy) {
    return orderBy.reduce((arr, value, i) => {
      arr.push(value)
      if (i < orderBy.length - 1) {
        arr.push(',')
      }

      return arr
    }, [])
  }

  _getWindowXpr (columns, orderBy) {
    const xpr = [{ func: 'ROW_NUMBER', args: [] }, 'OVER', '(', 'PARTITION BY', ...columns]
    if (orderBy.length !== 0) {
      xpr.push('ORDER BY', ...this._getOrderByForWindowFn(orderBy))
    }
    xpr.push(')')

    return { xpr: xpr, as: 'rowNumber' }
  }

  _copyColumns (columns, alias) {
    return columns.map(element => {
      const column = {
        ref: [alias, element.ref[element.ref.length - 1]]
      }

      if (element.as) {
        column.as = element.as
      }

      return column
    })
  }

  _convertOnToWhere (on, currentAlias, newAlias) {
    return on.map(element => {
      if (typeof element === 'object') {
        return {
          ref: [element.ref[0] === currentAlias ? newAlias : element.ref[0], element.ref[1]]
        }
      }

      return element
    })
  }

  _copyWhere (list) {
    return list.map(entry => {
      return typeof entry === 'object' ? this._copyObject(entry) : entry
    })
  }

  _copyObject (obj) {
    const newObj = {}

    for (const key of Object.keys(obj)) {
      if (Array.isArray(obj[key])) {
        newObj[key] = Array.from(obj[key])
      } else {
        newObj[key] = obj[key]
      }
    }

    return newObj
  }

  /**
   * Reduce column list to column(s) needed to merge the result into one.
   * @returns {Object}
   * @private
   */
  _getFilterExpandCQN (readToOneCQN, on, parentAlias, keyObject) {
    const columns = []

    const outerColumns = []

    for (const entry of on) {
      if (typeof entry === 'object' && entry.ref[0] === 'filterExpand') {
        columns.push(this._getColumnObjectForFilterExpand(readToOneCQN, parentAlias, entry.ref[1]))
        outerColumns.push({ ref: [entry.ref[1]] })
      }
    }

    const keys = Object.keys(keyObject).filter(key => key !== 'IsActiveEntity')

    for (const key of keys) {
      if (!columns.map(entry => entry.as).includes(key)) {
        columns.push(this._getColumnObjectForFilterExpand(readToOneCQN, parentAlias, key))
        outerColumns.push({ ref: [key] })
      }
    }

    const subSelect = Object.assign({}, readToOneCQN, { columns })

    const SELECT = { from: { SELECT: subSelect }, columns: outerColumns, distinct: true }

    return {
      SELECT: SELECT,
      as: 'filterExpand'
    }
  }

  _getColumnObjectForFilterExpand (readToOneCQN, parentAlias, key) {
    const relevantColumn = readToOneCQN.columns.find(
      column => column[IDENTIFIER] === key && column.ref[0] === parentAlias
    )
    return {
      ref: [parentAlias, (relevantColumn && relevantColumn.as) || key],
      as: key
    }
  }

  _getValueFromEntry (entry, parentAlias, key) {
    let value = entry[key] || entry[key.toUpperCase()]
    if (value === undefined) {
      value = entry[`${parentAlias}_${key}`] || entry[`${parentAlias}_${key}`.toUpperCase()]
    }
    return value
  }

  /**
   * In case a column is used at a JOIN, it needs to be added to the list of selected columns.
   * @returns {Array}
   * @private
   */
  _getJoinColumnsFromOnAddToMapping (mapping, parentAlias, on, entity) {
    const columns = []
    const columnNames = []

    for (const keyName of Object.keys(entity.keys)) {
      const columnNameAlt = keyName === 'IsActiveEntity' ? 'IsActiveEntity' : `${parentAlias}_${keyName}`
      if (!columnNames.includes(columnNameAlt)) {
        columnNames.push(columnNameAlt)
      }
    }

    for (const entry of on) {
      if (typeof entry === 'object' && entry.ref[0] !== 'filterExpand') {
        const as = entry.ref.join('_')
        columns.push({
          ref: entry.ref,
          as: as,
          [SKIP_MAPPING]: true
        })
      }
    }

    // Function will be used a post processing to create unique keys for cache and lookup the same
    mapping[GET_KEY_VALUE] = (atExpanded, entry) => {
      const keyValue = []
      const keyList = atExpanded
        ? Object.keys(entry).filter(keyName => keyName.toLowerCase().startsWith('filterexpand_'))
        : columnNames

      for (const key of keyList) {
        keyValue.push(this._getValueFromEntry(entry, parentAlias, key))
      }

      return keyValue.join(':')
    }

    return columns
  }

  /**
   * Get the explicitly named columns for expand and add ID columns, so the result can be added to the correct part at merged result.
   * @returns {Object}
   * @private
   */
  _getColumnsForExpand ({ tableAlias, columnList, entity, joinColumns, isActive, parentEntity = {} }) {
    const columns = []
    const keys = this._getKeyNames(entity)
    const parentKeys = this._getKeyNames(parentEntity)

    for (const column of columnList.expand) {
      if (column.expand || !column.ref) {
        columns.push(column)
      } else {
        this._addToColumnList(columns, entity, tableAlias, column, isActive)
      }
    }

    this._addMissingJoinElements(columns, joinColumns)
    this._addMissingKeyColumns(columns, tableAlias, keys, isActive)
    this._addMissingParentKeyColumns(columns, 'filterExpand', parentKeys, isActive)

    return columns
  }

  _createCalculatedBooleanColumn (alias, isActive) {
    return {
      val: isActive,
      as: alias,
      cast: { type: 'cds.Boolean' }
    }
  }

  _createIsActiveEntityOfParent (isActive, tableAlias) {
    return {
      val: isActive,
      as: `${tableAlias}_IsActiveEntity`,
      cast: { type: 'cds.Boolean' },
      [SKIP_MAPPING]: true
    }
  }

  _addToColumnList (columns, entity, tableAlias, column, isActive) {
    const columnName = column.ref[column.ref.length - 1]

    if (typeof isActive === 'boolean') {
      if (columnName === 'IsActiveEntity') {
        columns.push(this._createCalculatedBooleanColumn('IsActiveEntity', isActive))

        return
      }

      if (columnName === 'HasActiveEntity') {
        columns.push(this._createCalculatedBooleanColumn('HasActiveEntity', false))

        return
      }

      if (isActive && columnName === 'HasDraftEntity') {
        columns.push(this._getHasDraftEntityXpr(entity, tableAlias, columns))

        return
      }
    }

    columns.push({
      ref: [tableAlias, columnName],
      as: column.as || `${tableAlias}_${columnName}`
    })
  }

  _isNotIncludedIn (columns) {
    return entry =>
      !columns.some(column => (column.ref && column.ref[1] === entry) || ('val' in column && column.as === entry))
  }

  /**
   * Add join columns if they are not already existing in the list.
   * @private
   */

  _addMissingJoinElements (columns, joinColumns, keys) {
    const isNotIncludedInColumns = this._isNotIncludedIn(columns)
    for (const joinColumn of joinColumns) {
      if (isNotIncludedInColumns(joinColumn.ref[1])) {
        columns.push(joinColumn)
      }
    }
  }

  /**
   * Add key columns if they are not already existing in the list.
   * @private
   */
  _addMissingKeyColumns (columns, tableAlias, keys, isActive) {
    for (const key of keys.filter(this._isNotIncludedIn(columns))) {
      if (key === 'IsActiveEntity' && typeof isActive === 'boolean') {
        columns.push(this._createCalculatedBooleanColumn(key, isActive))
      } else {
        columns.push({
          ref: [tableAlias, key],
          as: `${tableAlias}_${key}`
        })
      }
    }
  }

  _addMissingParentKeyColumns (columns, tableAlias, keys, parentIsActive) {
    for (const key of keys) {
      if (key === 'IsActiveEntity' && typeof parentIsActive === 'boolean') {
        columns.push(this._createIsActiveEntityOfParent(parentIsActive, tableAlias))
      } else {
        columns.push({
          ref: [tableAlias, key],
          as: `${tableAlias}_${key}`,
          [SKIP_MAPPING]: true
        })
      }
    }
  }
}

/**
 * Creates CQN(s) by using JOIN for all expanded entries, as expanding is not supported by SQL.
 * @param {Object} cqn - CQN with expanded columns
 * @param {Object} csn - Services CSN
 * @param {Boolean} useWindow - If sub select or window function should be used for top/skip at expand
 * @returns {Object}
 * @private
 */
const createJoinCQNFromExpanded = (cqn, csn, useWindow, locale) => {
  return new JoinCQNFromExpanded(cqn, csn, useWindow, locale).buildJoinQueries()
}

/**
 * Check if the given CQN is of type select and contains expand.
 * @param {Object} cqn
 * @returns {boolean}
 * @private
 */
const hasExpand = cqn => {
  if (cqn && cqn.SELECT && Array.isArray(cqn.SELECT.columns)) {
    return cqn.SELECT.columns.some(column => typeof column.expand === 'object')
  }

  return false
}

module.exports = {
  createJoinCQNFromExpanded,
  hasExpand
}
