// @sap-cloud-sdk/core/dist/request-builder/response-transformers.js
"use strict";
/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */
Object.defineProperty(exports, "__esModule", { value: true });
var entity_deserializer_1 = require("../entity-deserializer");
function transformReturnValueForUndefined(data, builderFn) {
    return builderFn(data);
}
exports.transformReturnValueForUndefined = transformReturnValueForUndefined;
function transformReturnValueForEntity(data, entityConstructor) {
    return entity_deserializer_1.deserializeEntity(data.d, entityConstructor).setOrInitializeRemoteState();
}
exports.transformReturnValueForEntity = transformReturnValueForEntity;
function transformReturnValueForEntityList(data, entityConstructor) {
    return data.d.results.map(function (entityJson) { return entity_deserializer_1.deserializeEntity(entityJson, entityConstructor).setOrInitializeRemoteState(); });
}
exports.transformReturnValueForEntityList = transformReturnValueForEntityList;
function transformReturnValueForComplexType(data, builderFn) {
    return builderFn(data.d);
}
exports.transformReturnValueForComplexType = transformReturnValueForComplexType;
function transformReturnValueForComplexTypeList(data, builderFn) {
    return data.d.results.map(function (json) { return builderFn(json); });
}
exports.transformReturnValueForComplexTypeList = transformReturnValueForComplexTypeList;
function transformReturnValueForEdmType(data, builderFn) {
    return builderFn(data.d);
}
exports.transformReturnValueForEdmType = transformReturnValueForEdmType;
function transformReturnValueForEdmTypeList(data, builderFn) {
    return data.d.results.map(builderFn);
}
exports.transformReturnValueForEdmTypeList = transformReturnValueForEdmTypeList;
