// @sap/cds-runtime/lib/cds-services/util/assert.js
const { all, resolve } = require('./thenable')
const getError = require('./getError')

const ISO_DATE_PART1 =
  '[1-9]\\d{3}-(?:(?:0[1-9]|1[0-2])-(?:0[1-9]|1\\d|2[0-8])|(?:0[13-9]|1[0-2])-(?:29|30)|(?:0[13578]|1[02])-31)'
const ISO_DATE_PART2 = '(?:[1-9]\\d(?:0[48]|[2468][048]|[13579][26])|(?:[2468][048]|[13579][26])00)-02-29'
const ISO_DATE = `(?:${ISO_DATE_PART1}|${ISO_DATE_PART2})`
const ISO_TIME_NO_MILLIS = '(?:[01]\\d|2[0-3]):[0-5]\\d:[0-5]\\d'
const ISO_TIME = `${ISO_TIME_NO_MILLIS}(?:\\.\\d{1,9})?`
const ISO_DATE_TIME = `${ISO_DATE}T${ISO_TIME_NO_MILLIS}(?:Z|[+-][01]\\d:?[0-5]\\d)`
const ISO_TIMESTAMP = `${ISO_DATE}T${ISO_TIME}(?:Z|[+-][01]\\d:?[0-5]\\d)`

const UUID_REGEX = /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i
const ISO_DATE_REGEX = new RegExp(`^${ISO_DATE}$`, 'i')
const ISO_TIME_REGEX = new RegExp(`^${ISO_TIME_NO_MILLIS}$`, 'i')
const ISO_DATE_TIME_REGEX = new RegExp(`^${ISO_DATE_TIME}$`, 'i')
const ISO_TIMESTAMP_REGEX = new RegExp(`^${ISO_TIMESTAMP}$`, 'i')

const ASSERT_VALID_ELEMENT = 'ASSERT_VALID_ELEMENT'
const ASSERT_RANGE = 'ASSERT_RANGE'
const ASSERT_FORMAT = 'ASSERT_FORMAT'
const ASSERT_DATA_TYPE = 'ASSERT_DATA_TYPE'
const ASSERT_ENUM = 'ASSERT_ENUM'
const ASSERT_NOT_NULL = 'ASSERT_NOT_NULL'
const ASSERT_REFERENCE_INTEGRITY = 'ASSERT_REFERENCE_INTEGRITY'
const ASSERT_DEEP_TO_ONE_ASSOCIATION = 'ASSERT_DEEP_TO_ONE_ASSOCIATION'
const ASSERT_DEEP_TO_MANY_ASSOCIATION = 'ASSERT_DEEP_TO_MANY_ASSOCIATION'

const AssertCodeText = {
  ASSERT_VALID_ELEMENT: e => `Element '${e}' is not valid`,
  ASSERT_RANGE: e => `Value of element '${e}' is not in specified range`,
  ASSERT_FORMAT: e => `Value of element '${e}' is not in specified format`,
  ASSERT_DATA_TYPE: e => `Value of element '${e}' is invalid according to type definition`,
  ASSERT_ENUM: e => `Value of element '${e}' is invalid according to enum declaration`,
  ASSERT_NOT_NULL: e => `Value of element '${e}' is required`,
  ASSERT_REFERENCE_INTEGRITY: e => `Reference integrity is violated for association '${e}'`,
  ASSERT_DEEP_TO_ONE_ASSOCIATION: e => `It is not allowed to modify sub documents in to-one Association '${e}'`,
  ASSERT_DEEP_TO_MANY_ASSOCIATION: e => `Deep insert is not allowed for to-many Association '${e}'`
}

const _enumValues = element => {
  return Object.keys(element).map(enumKey => {
    const enum_ = element[enumKey]
    const enumValue = enum_ && (enum_.val || enum_.value) // OLD CSN
    return enumValue ? enumValue['='] || enumValue : enumKey
  })
}

const _assertError = (code, entity, element, value, key) => {
  const { name, type, precision, scale } = element

  const error = getError(400, { code, message: AssertCodeText[code](name || key) })
  Object.assign(error, {
    entity: entity.name,
    element: name,
    type: element.items ? element.items.type : type,
    value
  })

  if (element.enum) {
    error.enum = _enumValues(element)
  }
  if (precision) {
    error.precision = precision
  }
  if (scale) {
    error.scale = scale
  }
  if (element.target) {
    error.target = element.target
  }
  return error
}

const _checkString = value => {
  return typeof value === 'string'
}

const _checkNumber = value => {
  return typeof value === 'number'
}

const _checkDecimal = (value, element) => {
  const [left, right] = String(value).split('.')
  return (
    _checkNumber(value) &&
    ((!element.precision || left.length <= element.precision - (element.scale || 0)) &&
      (!element.scale || ((right || '').length <= element.scale && parseFloat(right) !== 0)))
  )
}

const _checkInteger = value => {
  return _checkNumber(value) && parseInt(value, 10) === value
}

const _checkBoolean = value => {
  return typeof value === 'boolean'
}

const _checkBuffer = value => {
  return Buffer.isBuffer(value)
}

const _checkUUID = value => {
  return _checkString(value) && UUID_REGEX.test(value)
}

const _checkISODate = value => {
  return _checkString(value) && ISO_DATE_REGEX.test(value)
}

const _checkISOTime = value => {
  return _checkString(value) && ISO_TIME_REGEX.test(value)
}

const _checkISODateTime = value => {
  return _checkString(value) && ISO_DATE_TIME_REGEX.test(value)
}

const _checkISOTimestamp = value => {
  return _checkString(value) && ISO_TIMESTAMP_REGEX.test(value)
}

const _checkInRange = (val, range) => {
  return _checkISODate(val)
    ? (new Date(val) - new Date(range[0])) * (new Date(val) - new Date(range[1])) <= 0
    : (val - range[0]) * (val - range[1]) <= 0
}

const _checkRegExpFormat = (val, format) => {
  return _checkString(val) ? val.match(new RegExp(format, 'g')) : false
}

const _isAssociationOrComposition = element => element.type === 'cds.Association' || element.type === 'cds.Composition'

const CDS_TYPE_CHECKS = {
  'cds.UUID': _checkUUID,
  'cds.Boolean': _checkBoolean,
  'cds.Integer': _checkInteger,
  'cds.Integer64': _checkInteger,
  'cds.Decimal': _checkDecimal,
  'cds.DecimalFloat': _checkNumber,
  'cds.Double': _checkNumber,
  'cds.Date': _checkISODate,
  'cds.Time': _checkISOTime,
  'cds.DateTime': _checkISODateTime,
  'cds.Timestamp': _checkISOTimestamp,
  'cds.String': _checkString,
  'cds.Binary': _checkBuffer,
  'cds.LargeString': _checkString,
  'cds.LargeBinary': _checkBuffer
}

// Limitation: depth 1
const checkComplexType = ([key, value], elements, ignoreNonModelledData) => {
  let found = false

  for (const objKey in elements) {
    if (objKey.startsWith(`${key}_`)) {
      const element = elements[objKey]
      const check = CDS_TYPE_CHECKS[element.type]
      found = true

      const nestedData = value[objKey.substring(key.length + 1)]
      // check existence of nestedData to not stumble across not-provided, yet-modelled type parts with depth > 1
      if (nestedData && !check(nestedData)) {
        return false
      }
    }
  }

  return found || ignoreNonModelledData
}

const _checkStaticElementByKey = (entity, key, value, result, ignoreNonModelledData) => {
  const element = entity.elements[key]

  if (!element) {
    if (!checkComplexType([key, value], entity.elements, ignoreNonModelledData)) {
      result.push(_assertError(ASSERT_VALID_ELEMENT, entity, { name: key }))
    }

    return result
  }

  const check = CDS_TYPE_CHECKS[element.type]

  if (check && !check(value, element)) {
    // code, entity, element, value
    result.push(_assertError(ASSERT_DATA_TYPE, entity, element, value, key))
  }

  return result
}

const checkInputConstraints = (entity, key, value, result) => {
  const element = entity.elements[key]

  if (!element) {
    return result
  }

  const enumElements = (element['@assert.range'] && element.enum) || element['@assert.enum'] ? element.enum : undefined
  const rangeElements = element['@assert.range'] && !enumElements ? element['@assert.range'] : undefined
  const formatElements = element['@assert.format']

  if (enumElements && !_enumValues(enumElements).includes(value)) {
    result.push(_assertError(ASSERT_ENUM, entity, element, value))
  }

  if (rangeElements && !_checkInRange(value, rangeElements)) {
    result.push(_assertError(ASSERT_RANGE, entity, element, value))
  }

  if (formatElements && !_checkRegExpFormat(value, formatElements)) {
    result.push(_assertError(ASSERT_FORMAT, entity, element, value))
  }

  return result
}

const checkStatic = (entity, data, ignoreNonModelledData = false) => {
  if (!Array.isArray(data)) {
    return checkStatic(entity, [data], ignoreNonModelledData)
  }

  return data.reduce((result, row) => {
    return Object.entries(row)
      .filter(([key, value]) => value !== null && value !== undefined)
      .reduce((result, [key, value]) => {
        return _checkStaticElementByKey(entity, key, value, result, ignoreNonModelledData)
      }, result)
  }, [])
}

const _isMandatoryField = element => {
  return (
    (element.notNull ||
      (element['@Common.FieldControl'] && element['@Common.FieldControl']['#'] === 'Mandatory') ||
      element['@Common.FieldControl.Mandatory'] ||
      element['@FieldControl.Mandatory'] ||
      element['@mandatory']) &&
    !element.default &&
    !_isAssociationOrComposition(element)
  )
}

const _isNotFilled = value => {
  return value === null || value === undefined || (typeof value === 'string' && value.trim() === '')
}

const checkNotNull = (entity, data) => {
  if (!Array.isArray(data)) {
    return checkNotNull(entity, [data])
  }

  return data.reduce((result, row) => {
    return Object.keys(row).reduce((result, key) => {
      const element = entity.elements[key]

      if (!element) {
        result.push(_assertError(ASSERT_VALID_ELEMENT, entity, { name: key }))
        return result
      }

      if (_isMandatoryField(element) && _isNotFilled(row[key])) {
        result.push(_assertError(ASSERT_NOT_NULL, entity, element, row[key]))
      }

      return result
    }, result)
  }, [])
}

const checkNotNullAll = (entity, data) => {
  if (!Array.isArray(data)) {
    return checkNotNullAll(entity, [data])
  }

  return Object.keys(entity.elements).reduce((result, key) => {
    const element = entity.elements[key]

    return data.reduce((result, row) => {
      if (_isMandatoryField(element) && _isNotFilled(row[key])) {
        result.push(_assertError(ASSERT_NOT_NULL, entity, element, row[key]))
      }

      return result
    }, result)
  }, [])
}

const _checkExistsWhere = (entity, whereList, context) => {
  const checks = whereList.map(where => {
    if (where.length === 0) {
      return true
    }

    const cqn = {
      SELECT: {
        columns: [{ val: 1, as: '_exists' }],
        from: { ref: [entity.name || entity] },
        where: where
      }
    }

    return context.run(cqn).then(exists => {
      return exists.length !== 0
    })
  })

  return all(checks)
}

const _checkExists = (entity, data, context) => {
  if (!Array.isArray(data)) {
    return _checkExists(entity, [data], context).then(result => {
      return result[0]
    })
  }

  const where = data.map(row => {
    return Object.keys(entity.keys).reduce((where, name) => {
      if (row[name] !== undefined && row[name] !== null) {
        if (where.length > 0) {
          where.push('and')
        }
        where.push({ ref: [name] }, '=', { val: row[name] })
      }

      return where
    }, [])
  })
  return _checkExistsWhere(entity, where, context)
}

const _getFullForeignKeyName = (elementName, foreignKeyName) => `${elementName}_${foreignKeyName}`

const _foreignKeyReducer = (key, foreignKeyName, row, element) => {
  const fullForeignKeyName = _getFullForeignKeyName(element.name, foreignKeyName)

  key[foreignKeyName] = row.hasOwnProperty(fullForeignKeyName) ? row[fullForeignKeyName] : null

  return key
}

const _buildForeignKey = (element, row) => {
  let foreignKey

  if (element.keys) {
    foreignKey = element.keys
      .map(obj => obj.ref[obj.ref.length - 1])
      .reduce((key, foreignKeyName) => {
        return _foreignKeyReducer(key, foreignKeyName, row, element)
      }, {})
  }

  return foreignKey
}

const _checkAssociations = (entity, element, row, context, result) => {
  if ((context.event === 'CREATE' || context.event === 'UPDATE') && row[element.name] !== undefined) {
    if (element.is2many) {
      result.push(_assertError(ASSERT_DEEP_TO_MANY_ASSOCIATION, entity, element))
      return result
    }

    if (element.is2one) {
      result.push(_assertError(ASSERT_DEEP_TO_ONE_ASSOCIATION, entity, element))
      return result
    }
  }
}

const _checkCreateUpdate = (result, key, entity, checks, data, context) => {
  const element = entity.elements[key]

  return data.reduce((result, row) => {
    const assocError = _checkAssociations(entity, element, row, context, result)
    if (assocError) {
      return assocError
    }

    // TODO: Why don't we check referential integrity for assoc with on condition?
    if (entity.elements[key].on) {
      return result
    }

    const foreignKey = _buildForeignKey(element, row)

    checks.push(
      _checkExists(element._target, foreignKey, context).then(exists => {
        if (!exists) {
          result.push(_assertError(ASSERT_REFERENCE_INTEGRITY, entity, element, foreignKey))
        }
      })
    )

    return result
  }, result)
}

const _isOnCondWith = (element, name) => {
  return element.some(el => {
    return el.ref && el.ref.includes(name)
  })
}

const _isCustomOnCond = element => {
  return element.on && !_isOnCondWith(element.on, '$self')
}

const _isManagedToOneAssoc = (element, name) => {
  return element.type === 'cds.Association' && element.is2one && element.target === name && !_isCustomOnCond(element)
}

const _buildWhereDelete = (result, key, element, context) => {
  return Object.keys(context.data).reduce((result, name) => {
    if (key.ref[0] === name) {
      if (result.length > 0) {
        result.push('and')
      }
      result.push({ ref: [_getFullForeignKeyName(element.name, key.ref[0])] }, '=', { val: context.data[name] })
    }

    return result
  }, result)
}

const _isCompBacklink = (entity, assoc) => {
  for (const element in entity.elements) {
    const targetElement = entity.elements[element]
    if (
      targetElement.type === 'cds.Composition' &&
      targetElement.on &&
      _isOnCondWith(targetElement.on, assoc.name) &&
      _isOnCondWith(targetElement.on, '$self')
    ) {
      return true
    }
  }
  return false
}
const _checkDelete = (result, key, entity, checks, context, csn) => {
  const elements = csn.definitions[key].elements
  const source = csn.definitions[key].name

  return Object.keys(elements).reduce((result, assoc) => {
    if (
      _isManagedToOneAssoc(elements[assoc], entity.name) &&
      csn.definitions[key].name !== entity.name &&
      elements[assoc]['@assert.integrity'] !== false
    ) {
      if (_isCompBacklink(entity, elements[assoc])) {
        return result
      }
      const keys = elements[assoc].keys
      // In case the other entity has no reference to the entity being deleted -> no dangling key
      if (!keys) return
      const where = keys.reduce((buildWhere, key) => {
        return _buildWhereDelete(buildWhere, key, elements[assoc], context)
      }, [])
      checks.push(
        _checkExistsWhere(source, [where], context).then(exists => {
          if (exists.includes(true)) {
            // console.log (source, assoc, '>', entity.name)
            result.push(_assertError(ASSERT_REFERENCE_INTEGRITY, entity, elements[assoc], context.data))
          }
        })
      )

      return result
    } else {
      return result
    }
  }, result)
}

const checkReferenceIntegrity = (entity, data, context, csn) => {
  if (!Array.isArray(data)) {
    return checkReferenceIntegrity(entity, [data], context, csn)
  }

  const checks = []
  let result
  if (context.event === 'CREATE' || context.event === 'UPDATE') {
    result = Object.keys(entity.elements)
      .filter(
        key =>
          entity.elements[key].type === 'cds.Association' &&
          entity.elements[key]._target['@cds.persistence.skip'] !== true &&
          entity.elements[key]['@assert.integrity'] !== false
      )
      .reduce((createUpdateResult, key) => {
        return _checkCreateUpdate(createUpdateResult, key, entity, checks, data, context)
      }, [])
  }
  if (context.event === 'DELETE') {
    // we are only interested in table-level references not all derived ones on view levels
    // TODO: why?
    while (entity.query && entity.query._target) {
      entity = csn.definitions[entity.query._target.name]
    }
    result = Object.keys(csn.definitions)
      .filter(key => csn.definitions[key].elements !== undefined)
      .reduce((deleteResult, key) => {
        return _checkDelete(deleteResult, key, entity, checks, context, csn)
      }, [])
  }

  if (checks.length) {
    return Promise.all(checks).then(() => {
      return result
    })
  }

  return resolve(result || [])
}

const checkKeys = (entity, data) => {
  if (!Array.isArray(data)) {
    return checkKeys(entity, [data])
  }

  const entityKeys = Object.keys(entity.keys)
  return data.reduce((result, row) => {
    for (const key of entityKeys) {
      if (entityKeys.some(key => row[key] === undefined)) {
        result.push(_assertError(ASSERT_NOT_NULL, entity, entity.elements[key]))
      }
    }
    return result
  }, [])
}

module.exports = {
  CDS_TYPE_CHECKS,
  checkComplexType,
  checkStatic,
  checkInputConstraints,
  checkKeys,
  checkNotNull,
  checkNotNullAll,
  checkReferenceIntegrity
}
