// @sap/cds-compiler/lib/api/options.js
'use strict';

const { validate, generateStringValidator } = require('./validate');

// Options that are advertised and documented to users
const publicOptionsNewAPI = [
  // GENERAL
  'beta',
  'longAutoexposed',
  'severities',
  'messages',
  // DB
  'sqlDialect',
  'sqlMapping',
  'magicVars',
  // ODATA
  'odataVersion',
  'odataFormat',
  'odataContainment',
  'odataProxies',
  'service',
];

// Internal options used for testing/debugging etc.
const privateOptions = [
  'lintMode',
  'fuzzyCsnError',
  'traceFs',
  'traceParser',
  'traceParserAmb',
  'hanaFlavor',
  'testMode',
  'compatibility',
];

const overallOptions = publicOptionsNewAPI.concat(privateOptions);

/**
 * Extract the cds relevant options from the provided options
 * Apply defaults and make sure that the "hard requirements" are met,
 * i.e. src: sql if to.sql() was called.
 *
 * @param {FlatOptions} [input={}] Input options
 * @param {FlatOptions} [defaults={}] Default options to apply
 * @param {FlatOptions} [hardRequire={}] Hard requirements to enforce
 * @param {object} [customValidators] Custom validations to run instead of defaults
 * @param {string[]} [combinationValidators] Option combinations to validate
 * @returns {TranslatedOptions} General cds options
 */
function translateOptions(input = {}, defaults = {}, hardRequire = {},
                          customValidators, combinationValidators) {
  const options = Object.assign({ messages: [] }, defaults);
  const inputOptionNames = Object.keys(input);
  for (const name of overallOptions) {
    // Ensure that arrays are not passed as a reference!
    // This caused issues with the way messages are handled in processMessages
    if (Array.isArray(input[name]) && inputOptionNames.indexOf(name) !== -1)
      options[name] = [ ...input[name] ];
    else if (inputOptionNames.indexOf(name) !== -1)
      options[name] = input[name];
  }
  // Validate the filtered input options
  // only "new-style" options are here
  validate(options, customValidators, combinationValidators);

  // Overwrite with the hardRequire options - like src: sql in to.sql()
  Object.assign(options, hardRequire);

  for (const optionName in options) {
    const optionValue = options[optionName];
    mapToOldNames(optionName, optionValue);
  }

  /**
   * Map a new-style option to it's old format
   *
   * @param {string} optionName Name of the option to map
   * @param {any} optionValue Value of the option to map
   */
  function mapToOldNames(optionName, optionValue) {
    // Keep all input options and add the "compatibility" options
    switch (optionName) {
      case 'beta':
        options.betaMode = optionValue;
        break;
      case 'odataVersion':
        options.version = optionValue;
        break;
      case 'sqlDialect':
        if (optionValue === 'hana') {
          options.dialect = optionValue;
        }
        else if (optionValue === 'sqlite') {
          options.dialect = optionValue;
          options.associations = 'joins';
        }
        else if (optionValue === 'hanaJoins') {
          options.dialect = 'hana';
          options.associations = 'joins';
        }
        else if (optionValue === 'hanaAssocs') {
          options.dialect = 'hana';
          options.associations = 'assocs';
        }
        break;
      case 'sqlMapping':
        options.names = optionValue;
        break;
      case 'magicVars':
        if (optionValue.user)
          options.user = optionValue.user;
        if (optionValue.locale)
          options.locale = optionValue.locale;
        break;
      default: break;
    }
  }

  return options;
}

module.exports = {
  to: {
    cdl: options => translateOptions(options),
    sql: (options) => {
      const hardOptions = { src: 'sql' };
      const defaultOptions = { sqlMapping: 'plain', sqlDialect: 'sqlite' };
      const processed = translateOptions(options, defaultOptions, hardOptions, undefined, [ 'sql-dialect-and-naming' ]);

      const result = Object.assign({}, processed);
      result.toSql = Object.assign({}, processed);

      return result;
    },
    hdi: (options) => {
      const hardOptions = { src: 'hdi' };
      const defaultOptions = { sqlMapping: 'plain', sqlDialect: 'hanaJoins' };
      const processed = translateOptions(options, defaultOptions, hardOptions, { sqlDialect: generateStringValidator([ 'hana', 'hanaJoins', 'hanaAssocs' ]) });
      const result = Object.assign({}, processed);
      result.toSql = Object.assign({}, processed);

      return result;
    },
    hdbcds: (options) => {
      const defaultOptions = { sqlMapping: 'plain' };
      const processed = translateOptions(options, defaultOptions, {}, { sqlDialect: generateStringValidator([ 'hana', 'hanaJoins', 'hanaAssocs' ]) });

      const result = Object.assign({}, processed);
      result.toHana = Object.assign({}, processed);


      return result;
    },
    edm: (options) => {
      const hardOptions = { json: true, combined: true };
      const defaultOptions = { odataVersion: 'v4', odataFormat: 'flat' };
      const processed = translateOptions(options, defaultOptions, hardOptions, { odataVersion: generateStringValidator([ 'v4' ]) }, [ 'valid-structured' ]);

      const result = Object.assign({}, processed);
      result.toOdata = Object.assign({}, processed);


      return result;
    },
    edmx: (options) => {
      const hardOptions = { xml: true, combined: true };
      const defaultOptions = {
        odataVersion: 'v4', odataFormat: 'flat',
      };
      const processed = translateOptions(options, defaultOptions, hardOptions, undefined, [ 'valid-structured' ]);

      const result = Object.assign({}, processed);
      result.toOdata = Object.assign({}, processed);

      return result;
    },
  },
  for: {   // TODO: Rename version to oDataVersion

    odata: (options) => {
      const defaultOptions = { odataVersion: 'v4', odataFormat: 'flat' };
      const processed = translateOptions(options, defaultOptions, undefined, undefined, [ 'valid-structured' ]);

      const result = Object.assign({}, processed);
      result.toOdata = Object.assign({}, processed);


      return result;
    },
    hana: (options) => {
      const defaultOptions = { sqlMapping: 'plain', sqlDialect: 'hanaAssocs' };
      const processed = translateOptions(options, defaultOptions);

      const result = Object.assign({}, processed);
      result.toHana = Object.assign({}, processed);


      return result;
    },
  },
};


/**
 * Flat input object using the new-style options.
 *
 * @typedef {object} FlatOptions
 */

/**
 * Flat options object, with defaults, validation and compatibility applied.
 *
 * @typedef {object} TranslatedOptions
 */
