// @sap/cds-runtime/lib/cds-services/adapter/odata-v4/odata-to-cqn/ExpressionToCQN.js
const odata = require('@sap/odata-server')
const { getOnCond } = require('../../../../../lib/cds-sql').onCond
const ExpressionKind = odata.uri.Expression.ExpressionKind
const BinaryOperatorKind = odata.uri.BinaryExpression.OperatorKind
const UnaryOperatorKind = odata.uri.UnaryExpression.OperatorKind
const MethodKind = odata.uri.MethodExpression.MethodKind
const ResourceKind = odata.uri.UriResource.ResourceKind
const EdmPrimitiveTypeKind = odata.edm.EdmPrimitiveTypeKind
const { getFeatureNotSupportedError } = require('../../../util/errors')

const _binaryOperatorToCQN = new Map([
  [BinaryOperatorKind.EQ, '='],
  [BinaryOperatorKind.NE, '!='],
  [BinaryOperatorKind.GE, '>='],
  [BinaryOperatorKind.GT, '>'],
  [BinaryOperatorKind.LE, '<='],
  [BinaryOperatorKind.LT, '<']
])

const toStrMethods = ['year', 'month', 'dayofmonth', 'second', 'hour', 'minute']

class ExpressionToCQN {
  constructor (entity, model, alias) {
    this._model = model
    this._entity = entity
    this._alias = alias
  }

  _convert (expression) {
    const type = expression.getType()

    if (!type) {
      return { val: null }
    }

    const value = expression.getText()

    if (value === null) return { val: null }

    switch (type) {
      case EdmPrimitiveTypeKind.Boolean:
        return { val: value === true || value === 'true' }
      case EdmPrimitiveTypeKind.Byte:
      case EdmPrimitiveTypeKind.SByte:
      case EdmPrimitiveTypeKind.Int16:
      case EdmPrimitiveTypeKind.Int32:
        return { val: parseInt(value) }
      case EdmPrimitiveTypeKind.Decimal:
      case EdmPrimitiveTypeKind.Single:
      case EdmPrimitiveTypeKind.Double:
        return { val: parseFloat(value) }
      default:
        return { val: value }
    }
  }

  _lambda (pathSegments) {
    const navName = this._segmentFromMember(pathSegments[0])
    const condition = this._segmentFromMember(pathSegments[1])
    const onCondition = getOnCond(this._entity.elements[navName], navName, this._model, 'foreign', 'parent')
    const parentKeys = onCondition.filter(e => e.ref && e.ref[0] === 'parent').map(e => e.ref[1])
    const foreignKeys = onCondition.filter(e => e.ref && e.ref[0] === 'foreign').map(e => e.ref[1])
    const lambdaOp = pathSegments[1].getKind() === ResourceKind.ANY_EXPRESSION ? 'any' : 'all'
    const subSelect = {
      SELECT: {
        columns: foreignKeys.map(fk => ({ ref: [fk] })),
        from: { ref: [this._entity.elements[navName].target] },
        where: condition
      }
    }

    return [{ list: parentKeys.map(pk => ({ ref: [pk] })) }, '=', this._genericFn(lambdaOp, [subSelect])]
  }

  _segmentFromMember (segment) {
    switch (segment.getKind()) {
      case ResourceKind.PRIMITIVE_PROPERTY:
      case ResourceKind.COMPLEX_PROPERTY:
      case ResourceKind.PRIMITIVE_COLLECTION_PROPERTY:
      case ResourceKind.COMPLEX_COLLECTION_PROPERTY:
        return segment.getProperty().getName()
      case ResourceKind.NAVIGATION_TO_ONE:
      case ResourceKind.NAVIGATION_TO_MANY:
        return segment.getNavigationProperty().getName()
      case ResourceKind.ALL_EXPRESSION:
      case ResourceKind.ANY_EXPRESSION:
        return this.parse(segment.getExpression())
      default:
        throw getFeatureNotSupportedError(`Segment kind "${segment.getKind()}" in $filter query option`)
    }
  }

  _getMemberRecursively (pathSegments) {
    const [segment, ...nextSegments] = pathSegments

    if (segment.getKind() === ResourceKind.NAVIGATION_TO_ONE) {
      return [this._segmentFromMember(segment), ...this._getMemberRecursively(nextSegments)]
    }

    if (segment.getKind() === ResourceKind.EXPRESSION_VARIABLE) {
      return [...this._getMemberRecursively(nextSegments)]
    }

    return [this._segmentFromMember(segment)]
  }

  _member (expression) {
    const pathSegments = expression.getPathSegments()
    if (pathSegments.some(segment => segment.getKind() === ResourceKind.NAVIGATION_TO_MANY)) {
      return this._lambda(pathSegments)
    }

    const aliasNeeded = this._alias && pathSegments[0].getKind() !== ResourceKind.EXPRESSION_VARIABLE
    const members = this._getMemberRecursively(pathSegments)
    return { ref: aliasNeeded ? [this._alias, ...members] : members }
  }

  _getParameters (expression) {
    return expression.getParameters().map(parameter => {
      return this.parse(parameter)
    })
  }

  _like (column, value, operator) {
    return [column, operator ? 'not like' : 'like', value]
  }

  _genericFn (methodName, args) {
    return { func: methodName, args }
  }

  /**
   * Evaluate an method expression, which in SQL would be 'column condition value'.
   * Can also be nested.
   * @param {Object} expression
   * @param {string} [operator] - Operator, that might be used to invert a method or similar
   * @throws Error - if method expression is not supported
   * @private
   */
  _method (expression, operator) {
    const parameters = this._getParameters(expression)
    switch (expression.getMethod()) {
      case MethodKind.CONTAINS:
        return this._like(parameters[0], { func: 'concat', args: ["'%'", parameters[1], "'%'"] }, operator)
      case MethodKind.STARTSWITH:
        return this._like(parameters[0], { func: 'concat', args: [parameters[1], "'%'"] }, operator)
      case MethodKind.ENDSWITH:
        return this._like(parameters[0], { func: 'concat', args: ["'%'", parameters[1]] }, operator)
      case MethodKind.NOW:
        return { val: new Date().toISOString() }
      case MethodKind.TOUPPER:
        return this._genericFn('upper', parameters)
      case MethodKind.TOLOWER:
        return this._genericFn('lower', parameters)
      case MethodKind.INDEXOF:
        return this._genericFn('locate', parameters)
      case MethodKind.SUBSTRING:
        // special case: odata function returns 0-based indexes
        // hana returns 1-based indexes
        parameters[1].val++
        return this._genericFn('substring', parameters)
      case MethodKind.DAY:
        return this._genericFn('dayofmonth', parameters)
      case MethodKind.DATE:
        return this._genericFn('to_date', parameters)
      case MethodKind.TIME:
        return this._genericFn('to_time', parameters)
      case MethodKind.CEILING:
        return this._genericFn('ceil', parameters)
      case MethodKind.TRIM:
      case MethodKind.LENGTH:
      case MethodKind.CONCAT:
      case MethodKind.HOUR:
      case MethodKind.MINUTE:
      case MethodKind.SECOND:
      case MethodKind.MONTH:
      case MethodKind.YEAR:
      case MethodKind.FLOOR:
      case MethodKind.ROUND:
        return this._genericFn(expression.getMethod(), parameters)
      default:
        throw getFeatureNotSupportedError(`Method "${expression.getMethod()}" in $filter or $orderby query options`)
    }
  }

  _fillAfterDot (val) {
    let [beforeDot, afterDot = ''] = val.split('.')

    while (afterDot.length < 3) {
      afterDot = afterDot.concat(0)
    }

    return `${beforeDot}.${afterDot}`
  }

  _convertValToString (valueObj) {
    return `${valueObj.val < 10 ? 0 : ''}${valueObj.val}`
  }

  _convertLocate (arg1, arg2) {
    if (arg1.func === 'locate') {
      arg2.val++
    }
  }

  _convertLocateForIndexOne (left, right) {
    this._convertLocate(left, right)
    this._convertLocate(right, left)
  }

  _convertDateFunctions (arg1, arg2) {
    if (arg1.func && toStrMethods.includes(arg1.func) && arg1.args) {
      arg2.val = this._convertValToString(arg2)
      if (arg1.func === 'second') {
        arg2.val = this._fillAfterDot(arg2.val)
      }
    }
  }

  _convertNumbersToStringForDateFunctions (left, right) {
    this._convertDateFunctions(left, right)
    this._convertDateFunctions(right, left)
  }

  _compare (operator, left, right) {
    // special case:
    // odata indexof function returns the zero-based character position of the first occurrence
    // hana locate function counts position beginning with 1, not 0
    this._convertLocateForIndexOne(left, right)

    // sqlite requires leading 0 for numbers, this works on hana as well so we generally add it
    this._convertNumbersToStringForDateFunctions(left, right)

    return [left, _binaryOperatorToCQN.get(operator), right]
  }

  _binary (expression) {
    const operator = expression.getOperator()
    const left = this.parse(expression.getLeftOperand())
    const right = this.parse(expression.getRightOperand())

    switch (operator) {
      case BinaryOperatorKind.AND:
        return [...left, 'and', ...right]
      case BinaryOperatorKind.OR:
        return ['(', ...left, 'or', ...right, ')']
      case BinaryOperatorKind.NE:
        return right && right.val === null
          ? [left, 'is not null']
          : ['(', left, '!=', right, 'or', left, 'is null', ')']
      case BinaryOperatorKind.EQ:
      case BinaryOperatorKind.GE:
      case BinaryOperatorKind.GT:
      case BinaryOperatorKind.LE:
      case BinaryOperatorKind.LT:
        return this._compare(operator, left, right)
      default:
        throw getFeatureNotSupportedError(`Binary operator "${expression.getOperator()}" in $filter query option`)
    }
  }

  _unary (expression) {
    if (expression.getOperator() !== UnaryOperatorKind.NOT) {
      throw getFeatureNotSupportedError(`Unary operator "${expression.getOperator()}" in $filter query option`)
    }

    return this.parse(expression.getOperand(), UnaryOperatorKind.NOT)
  }

  /**
   * Convert a odata-v4 filter expression object into a CQN object.
   *
   * @param {Object} expression - odata filter expression
   * @param {string} [operator] - Operator, that might be used to invert a method or similar
   * @throws Error - if expression object is not supported
   * @private
   */
  parse (expression, operator) {
    switch (expression.getKind()) {
      case ExpressionKind.ALIAS:
        return this.parse(expression.getExpression())
      case ExpressionKind.BINARY:
        return this._binary(expression)
      case ExpressionKind.LITERAL:
        return this._convert(expression)
      case ExpressionKind.MEMBER:
        return this._member(expression)
      case ExpressionKind.METHOD:
        return this._method(expression, operator)
      case ExpressionKind.UNARY:
        return this._unary(expression)
      default:
        throw getFeatureNotSupportedError(`Expression "${expression.getKind()}" in $filter or $orderby query options`)
    }
  }
}

module.exports = ExpressionToCQN
