// @sap/odata-server/lib/batch/BatchContext.js
'use strict';

const BatchErrorInfo = require('./BatchErrorInfo');

/**
 * Stores information required for batch processing, similar to the Context class.
 */
class BatchContext {
    /**
     * Constructor
     * @param {OdataRequest} incomingODataRequest Request received from the client
     * @param {OdataResponse} incomingODataResponse Response to be send to the client
     */
    constructor(incomingODataRequest, incomingODataResponse) {
        /**
         * OdataRequest wrapping the network request received by node.js
         * @type {OdataRequest}
         * @private
         */
        this._incomingODataRequest = incomingODataRequest;

        /**
         * OdataResponse wrapping the network response to be sent by node.js
         * @type {OdataResponse}
         * @private
         */
        this._incomingODataResponse = incomingODataResponse;

        /**
         * Requests extracted from the batch request's payload
         * @type {OdataRequestInBatch[]}
         * @private
         */
        this._requests = [];

        /**
         * Responses to be written into the batch response as part of the payload, with the request ID as key
         * @type {Map.<string, OdataResponseInBatch>}
         * @private
         */
        this._responses = new Map();

        /**
         * Map containing the location headers of the executed requests identified by the request ID,
         * used for content-ID referencing
         * @type {Map.<string, string>}
         * @private
         */
        this._locations = new Map();

        /**
         * Map containing for each atomicity group (identified by the atomicity-group ID) a list of BatchErrorInfo
         * instances to store which requests of the atomicity group have not been executed successfully.
         * Failed requests outside of an atomicity group are stored under the key <code>null</code>.
         * @type {Map.<?string, BatchErrorInfo[]>}
         * @private
         */
        this._failedRequestsPerAtomicityGroup = new Map();

        /**
         * Map containing the error object, thrown by a handler function or due to a framework error,
         * causing an atomicity group to fail
         * @type {Map.<string, ?Error>}
         * @private
         */
        this._failedAtomicityGroups = new Map();

        /**
         * Flag indicating if the batch processing should be continued after the first error
         * @type {boolean}
         * @private
         */
        this._continueOnError = this._incomingODataRequest.getPreferences().getOdataContinueOnError();

        /**
         * Boundary of the incoming batch request if multipart/mixed format is used. Used to create the response boundary.
         * @type {string}
         * @private
         */
        this._boundary = '';

        /**
         * Semantics of batch request
         * @type {BatchContext.SEMANTICS}
         * @private
         */
        this._semantics = BatchContext.SEMANTICS.MULTIPART;
    }

    /**
     * Returns the batch request received from the client.
     * @returns {OdataRequest} the batch request
     */
    getRequest() {
        return this._incomingODataRequest;
    }

    /**
     * Returns the batch response to be sent to the client.
     * @returns {OdataResponse} the batch response
     */
    getResponse() {
        return this._incomingODataResponse;
    }

    /**
     * Adds an OData response to the responses list for this batch request.
     * @param {OdataResponseInBatch} odataResponse Response
     * @returns {BatchContext} this instance
     */
    addResponseInBatch(odataResponse) {
        this._responses.set(odataResponse.getOdataResponseId(), odataResponse);
        return this;
    }

    /**
     * If a batched request has been processed and its resulting status code is between
     * 400 (inclusive) and 600 (exclusive), this method is called to store the error information.
     * The request is marked as failed and if the request belongs to an atomicity group
     * this group is also marked as failed.
     * @param {string} id ID of the batched request where the error occurred
     * @param {OdataRequestInBatch} request the batched request whose processing failed
     * @param {OdataResponseInBatch} response response object of the batched request whose processing failed
     * @returns {BatchContext} this instance
     */
    markRequestAsFailed(id, request, response) {
        const atomicityGroupId = request.getAtomicityGroupId();
        const batchErrorInfo = new BatchErrorInfo(id, request, response);

        if (this._failedRequestsPerAtomicityGroup.has(atomicityGroupId)) {
            this._failedRequestsPerAtomicityGroup.get(atomicityGroupId).push(batchErrorInfo);
        } else {
            this._failedRequestsPerAtomicityGroup.set(atomicityGroupId, [batchErrorInfo]);
        }

        // Also mark the whole atomicity group as failed, if the group is not already marked as failed
        // (e.g., due to an error in the previous batched request).
        if (!this._failedAtomicityGroups.get(atomicityGroupId)) this.markAtomicityGroupAsFailed(atomicityGroupId);
        return this;
    }

    /**
     * Mark an atomicity group as failed.
     * @param {string} atomicityGroupId atomicity group ID
     * @param {Error} [error] error information
     * @returns {BatchContext} this instance
     */
    markAtomicityGroupAsFailed(atomicityGroupId, error) {
        this._failedAtomicityGroups.set(atomicityGroupId, error);
        return this;
    }

    /**
     * Returns the map with failed atomicity groups.
     * @returns {Map.<string, ?Error>} the failed groups
     */
    getFailedAtomicityGroups() {
        return this._failedAtomicityGroups;
    }

    /**
     * Returns the failed requests of an atomicity group.
     * @param {string} atomicityGroupId the ID of the atomicity group
     * @returns {BatchErrorInfo[]} Erroneous requests
     */
    getFailedRequestsOfAtomicityGroup(atomicityGroupId) {
        return this._failedRequestsPerAtomicityGroup.get(atomicityGroupId) || [];
    }

    /**
     * Check if a atomicity group has been executed.
     * @param {string} id atomicity-group ID
     * @returns {boolean} whether the atomicity group has been executed
     */
    isAtomicityGroupExecuted(id) {
        return this._failedRequestsPerAtomicityGroup.get(id) === undefined;
    }

    /**
     * Check if a request has been executed and failed.
     * @param {string} id request ID
     * @returns {boolean} whether the request has been executed and failed
     */
    isRequestFailed(id) {
        const request = this._requests.find(req => req.getOdataRequestId() === id);
        if (!request) return false;
        const atomicityGroup = request.getAtomicityGroupId();
        return this._failedRequestsPerAtomicityGroup.has(atomicityGroup)
            && this._failedRequestsPerAtomicityGroup.get(atomicityGroup).some(info => info.getId() === id);
    }

    /**
     * Sets the list of OData requests which have been read from the batch request.
     * @param {OdataRequestInBatch[]} requestsInBatch requests collected by the batch-request parsing
     * @returns {BatchContext} this instance
     */
    setRequestList(requestsInBatch) {
        this._requests = requestsInBatch;
        return this;
    }

    /**
     * Returns the list of OData requests which have been read from the batch request.
     * @returns {OdataRequestInBatch[]} the request list
     */
    getRequestList() {
        return this._requests;
    }

    /**
     * Returns the OData responses for the batch response.
     * @returns {Map.<string, OdataResponseInBatch>} the responses, with the request ID as key
     */
    getResponses() {
        return this._responses;
    }

    /**
     * Returns whether the batch-request execution should be continued after an error.
     * @returns {boolean} whether the batch request execution should be continued after an error
     */
    isContinueOnError() {
        return this._continueOnError;
    }

    /**
     * Set the boundary of the incoming batch request. Used to create the response boundary.
     * @param {string} boundary the boundary
     * @returns {BatchContext} this instance
     */
    setBatchBoundary(boundary) {
        this._boundary = boundary;
        return this;
    }

    /**
     * Returns the boundary of the incoming batch request. Used to create the response boundary.
     * @returns {string} boundary
     */
    getBoundary() {
        return this._boundary;
    }

    /**
     * Set the semantics for batch processing.
     * @param {BatchContext.SEMANTICS} semantics the semantics
     * @returns {BatchContext} this instance
     */
    setSemantics(semantics) {
        this._semantics = semantics;
        return this;
    }

    /**
     * Returns the semantics for batch processing.
     * @returns {BatchContext.SEMANTICS} semantics
     */
    getSemantics() {
        return this._semantics;
    }

    /**
     * Sets a location from a CREATED response.
     * @param {string} id request ID
     * @param {string} location location of the created entity
     * @returns {BatchContext} this instance
     */
    addLocation(id, location) {
        this._locations.set(id, location);
        return this;
    }

    /**
     * Returns a location from a CREATED response.
     * @param {string} id the ID
     * @returns {string} the location
     */
    getLocation(id) {
        return this._locations.get(id);
    }

    /**
     * Remove an atomicity group.
     * @param {string} atomicityGroupId the ID of the atomicity group
     * @returns {BatchContext} this instance
     */
    removeAtomicityGroup(atomicityGroupId) {
        for (const request of this._requests) {
            if (request.getAtomicityGroupId() === atomicityGroupId) {
                const id = request.getOdataRequestId();
                this._responses.delete(id);
            }
        }

        // Remove atomicity group.
        this._failedRequestsPerAtomicityGroup.delete(atomicityGroupId);
        this._failedAtomicityGroups.delete(atomicityGroupId);

        return this;
    }
}


/**
 * Semantics of a batch request
 * MULTIPART means the Odata 4.0 batch semantics for multipart/mixed content type is used
 * JSON means the Odata 4.01 batch semantics for batch in application/json content type is used
 * @enum {string}
 * @type {{MULTIPART: string, JSON: string}}
 */
BatchContext.SEMANTICS = {
    MULTIPART: 'MULTIPART',
    JSON: 'JSON'
};

module.exports = BatchContext;
