// @sap/cds-runtime/lib/cds-services/adapter/auth/Mock.js
const DEFAULTS = {
  userAttributes: {},
  scopes: ['openid'],
  grantType: 'authorization_code',
  config: { identityzone: '', xsappname: '' },
  subdomain: '',
  xsappname: '',
  additionalAuthAttributes: '',
  serviceinstanceid: null
}

class Mock {
  constructor (users = {}) {
    if (typeof users === 'function') {
      this._validate = users
    } else {
      this._users = users
    }

    this._challenge = `Basic realm="Users"`
    this.name = 'mock'
  }

  authenticate (req) {
    let context = this._getContext(req)

    if (!context) {
      return context
    }

    // passport will set these in req.user & req.authInfo respectively
    if (context.then) {
      return context
        .then(context => {
          context = this._ensureXssecFormat(context)
          this.success(this._contextToUser(context), context)
        })
        .catch(() => {
          return this.fail(this._challenge)
        })
    }

    context = this._ensureXssecFormat(context)

    this.success(this._contextToUser(context), context)
  }

  _getContext (req) {
    const credentials = this._credentialsFromReq(req)

    if (!credentials) {
      return credentials
    }

    // Validation function provided instead of configuration
    if (this._validate) {
      return this._validate(credentials.userId, credentials.password, req)
    }

    const user = this._userFromUserId(credentials.userId)
    // Likely implicit auto wired setup
    if (Object.keys(this._users).length === 0 || (!user && this._users['*'] === true)) {
      return this._getMockedContext(credentials.userId)
    }

    // Explicit configuration
    if (this._credentialsDoNotMatch(user, credentials.userId, credentials.password)) {
      return this.fail(this._challenge)
    }

    return user.jwt || user
  }

  _credentialsFromReq (req) {
    const authorization = req.headers.authorization

    if (!authorization) {
      return this.fail(this._challenge)
    }

    const [scheme, base64] = authorization.split(' ')

    if (!scheme || scheme.toLowerCase() !== 'basic') {
      return this.fail(this._challenge)
    }

    if (!base64) {
      return this.fail(400)
    }

    const [userId, password] = Buffer.from(base64, 'base64')
      .toString()
      .split(':')

    return { userId, password }
  }

  _userFromUserId (userId) {
    if (this._users[userId] && !this._users[userId].ID) {
      return { ...this._users[userId], ID: userId }
    }
    for (const user in this._users) {
      if (this._users[user].ID === userId) {
        return this._users[user]
      }
    }
  }

  _credentialsDoNotMatch (user, userId, password) {
    if (!userId || !user) {
      return true
    }

    const givenPassword = password || ''
    const storedPassword = user.password || ''

    return storedPassword !== givenPassword
  }

  _getMockedContext (userId) {
    return {
      userInfo: {
        logonName: userId,
        givenName: userId,
        familyName: userId,
        email: `${userId}@acme.corp`
      }
    }
  }

  _contextToUser (context) {
    // Mimic @sap/xssec behavior
    if (!context.userInfo || !context.userInfo.logonName) {
      return {}
    }

    return {
      id: context.userInfo.logonName,
      name: {
        givenName: context.userInfo.givenName,
        familyName: context.userInfo.familyName
      },
      emails: [{ value: context.userInfo.email }]
    }
  }

  _ensureXssecFormat (context) {
    if (context.userInfo || context.grantType || context.xsappname) {
      return this._xssecDefaults(context)
    }

    return this._xssecDefaults(this._plainJwtToXssec(context))
  }

  _xssecDefaults (context) {
    this._extendWithDefaults(DEFAULTS, context)

    return context
  }

  _extendWithDefaults (defaults, object) {
    for (const key of Object.keys(defaults)) {
      if (object[key] === undefined) {
        object[key] = Array.isArray(defaults[key]) ? defaults[key].slice(0) : defaults[key]
      } else if (Array.isArray(object[key]) && Array.isArray(defaults[key])) {
        object[key].push(...defaults[key])
      } else if (typeof defaults[key] === 'object' && !Array.isArray(defaults[key]) && defaults[key] !== null) {
        this._extendWithDefaults(defaults[key], object[key])
      }
    }
  }

  _plainJwtToXssec (context) {
    const xsappname = Array.isArray(context.aud) ? context.aud[0] : undefined
    const identityzone = context.ext_attr ? context.ext_attr.zdn : undefined

    return {
      userAttributes: context['userAttributes'] || context['xs.user.attributes'] || context.attributes,
      scopes: context.scope || context.roles,
      userInfo: {
        logonName: context.user_name || context.ID,
        givenName: context.given_name,
        familyName: context.family_name,
        email: context.email
      },
      grantType: context.grant_type,
      config: {
        identityzone: identityzone,
        identityzoneid: context.zid,
        tenantid: context.zid,
        xsappname: xsappname
      },
      xsappname: xsappname,
      clientId: context.client_id,
      identityZone: context.zid,
      subdomain: identityzone,
      origin: context.origin
    }
  }
}

module.exports = Mock
