// @sap/odata-server/index.js
'use strict';

const commons = require('@sap/odata-commons');

module.exports = {
    FullQualifiedName: commons.FullQualifiedName,
    edm: {
        EdmTypeKind: commons.edm.EdmType.TypeKind,
        EdmPrimitiveTypeKind: commons.edm.EdmPrimitiveTypeKind
    },
    uri: {
        UriResource: commons.uri.UriResource,
        Expression: commons.uri.Expression,
        BinaryExpression: commons.uri.BinaryExpression,
        UnaryExpression: commons.uri.UnaryExpression,
        MethodExpression: commons.uri.MethodExpression,
        Transformation: commons.uri.apply.Transformation,
        BottomTopTransformation: commons.uri.apply.BottomTopTransformation,
        AggregateExpression: commons.uri.apply.AggregateExpression
    },
    QueryOptions: commons.uri.UriInfo.QueryOptions,
    HttpMethods: commons.http.HttpMethod.Methods,
    HttpStatusCodes: commons.http.HttpStatusCode.StatusCodes,
    PreferenceReturnValues: commons.http.Preferences.ReturnValues,
    ContentTypes: commons.format.ContentTypeInfo.ContentTypes,
    RepresentationKinds: commons.format.RepresentationKind.Kinds,

    ServiceFactory: require('./lib/ServiceFactory'),  // eslint-disable-line global-require
    ApplicationError: require('./lib/errors/ApplicationError'),  // eslint-disable-line global-require
    Components: require('./lib/core/ComponentManager').Components,  // eslint-disable-line global-require
    BatchExitHandler: require('./lib/batch/BatchExitHandler')  // eslint-disable-line global-require
};
