// @sap-cloud-sdk/core/dist/selectable/complex-type-field.js
"use strict";
/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var field_1 = require("./field");
/**
 * Represents a property of an OData entity with a complex type.
 *
 * `ComplexTypeField`s are used as static properties of entities and are generated from the metadata, i.e. for each property of
 * an OData entity, that has a complex type, there exists one static instance of `ComplexTypeField` (or rather one of its subclasses) in the corresponding generated class file.
 * `ComplexTypeField`s are used to represent the domain of complex or custom structures that can be used in select, filter and order by functions.
 * For example, when constructing a query on the TimeSheetEntry entity, an instance of `ComplexTypeField<TimeSheetEntry>`
 * can be supplied as argument to the select function, e.g. `TimeSheetEntry.TIME_SHEET_DATA_FIELDS`.
 * Moreover, classes implementing this abstract class will provide property fields, that can be used for filtering and ordering.
 *
 * See also: [[Selectable]]
 *
 * @typeparam EntityT - Type of the entity the field belongs to
 */
var ComplexTypeField = /** @class */ (function (_super) {
    __extends(ComplexTypeField, _super);
    /*
     * Union of the two possible constructors.
     */
    function ComplexTypeField(fieldName, fieldOf, complexTypeName) {
        var _this = _super.call(this, fieldName, getEntityConstructor(fieldOf)) || this;
        _this.fieldOf = fieldOf;
        _this.complexTypeName = complexTypeName;
        return _this;
    }
    ComplexTypeField.prototype.fieldPath = function () {
        var value = this.fieldOf instanceof ComplexTypeField ? this.fieldOf.fieldPath() + "/" + this._fieldName : this._fieldName;
        return value;
    };
    return ComplexTypeField;
}(field_1.Field));
exports.ComplexTypeField = ComplexTypeField;
/**
 * Convenience method to return the entity constructor in the complex extensions of the normal fields e.g. ComplexTypeStringPropertyField
 * @param arg - Contains either the entity containing the complex field or a complex field in case of nested fields.
 * @returns Constructable
 */
function getEntityConstructor(arg) {
    return arg instanceof ComplexTypeField ? arg._entityConstructor : arg;
}
exports.getEntityConstructor = getEntityConstructor;
/**
 * Convenience method to return the EDM type for the overloaed constructor e.g. ComplexTypeStringPropertyField
 * @param arg1 - Contains either the type name or the EdmType
 * @param arg2 - Contains either the EdmType or undefined
 * @returns EdmType
 */
function getEdmType(arg1, arg2) {
    if (arg1.includes('Edm.') && !arg2) {
        return arg1;
    }
    if (typeof arg1 === 'string' && arg2 && arg2.includes('Edm.')) {
        return arg2;
    }
    throw new Error('Illegal argument exception!');
}
exports.getEdmType = getEdmType;
