// @sap-cloud-sdk/core/dist/selectable/custom-field.js
"use strict";
/* Copyright (c) 2020 SAP SE or an SAP affiliate company. All rights reserved. */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var big_number_field_1 = require("./big-number-field");
var boolean_field_1 = require("./boolean-field");
var date_field_1 = require("./date-field");
var field_1 = require("./field");
var number_field_1 = require("./number-field");
var string_field_1 = require("./string-field");
var time_field_1 = require("./time-field");
var CustomField = /** @class */ (function (_super) {
    __extends(CustomField, _super);
    function CustomField(_fieldName, _entityConstructor) {
        var _this = _super.call(this, _fieldName, _entityConstructor) || this;
        _this._fieldName = _fieldName;
        _this._entityConstructor = _entityConstructor;
        return _this;
    }
    CustomField.prototype.edmString = function () {
        return new string_field_1.StringField(this._fieldName, this._entityConstructor, 'Edm.String');
    };
    CustomField.prototype.edmBoolean = function () {
        return new boolean_field_1.BooleanField(this._fieldName, this._entityConstructor, 'Edm.Boolean');
    };
    CustomField.prototype.edmGuid = function () {
        return new string_field_1.StringField(this._fieldName, this._entityConstructor, 'Edm.Guid');
    };
    CustomField.prototype.edmDecimal = function () {
        return new big_number_field_1.BigNumberField(this._fieldName, this._entityConstructor, 'Edm.Decimal');
    };
    CustomField.prototype.edmInt16 = function () {
        return new number_field_1.NumberField(this._fieldName, this._entityConstructor, 'Edm.Int16');
    };
    CustomField.prototype.edmInt32 = function () {
        return new number_field_1.NumberField(this._fieldName, this._entityConstructor, 'Edm.Int32');
    };
    CustomField.prototype.edmInt64 = function () {
        return new big_number_field_1.BigNumberField(this._fieldName, this._entityConstructor, 'Edm.Int64');
    };
    CustomField.prototype.edmSingle = function () {
        return new number_field_1.NumberField(this._fieldName, this._entityConstructor, 'Edm.Single');
    };
    CustomField.prototype.edmDouble = function () {
        return new number_field_1.NumberField(this._fieldName, this._entityConstructor, 'Edm.Double');
    };
    CustomField.prototype.edmByte = function () {
        return new number_field_1.NumberField(this._fieldName, this._entityConstructor, 'Edm.Byte');
    };
    CustomField.prototype.edmSByte = function () {
        return new number_field_1.NumberField(this._fieldName, this._entityConstructor, 'Edm.SByte');
    };
    CustomField.prototype.edmDateTime = function () {
        return new date_field_1.DateField(this._fieldName, this._entityConstructor, 'Edm.DateTime');
    };
    CustomField.prototype.edmDateTimeOffset = function () {
        return new date_field_1.DateField(this._fieldName, this._entityConstructor, 'Edm.DateTimeOffset');
    };
    CustomField.prototype.edmBinary = function () {
        return new string_field_1.StringField(this._fieldName, this._entityConstructor, 'Edm.Binary');
    };
    CustomField.prototype.edmTime = function () {
        return new time_field_1.TimeField(this._fieldName, this._entityConstructor, 'Edm.Time');
    };
    return CustomField;
}(field_1.Field));
exports.CustomField = CustomField;
